--- a/cmake/modules/FindPyKDE4.py.orig	2019-10-05 18:46:43.123448485 +0200
+++ b/cmake/modules/FindPyKDE4.py	2019-10-05 18:49:02.073443736 +0200
@@ -2,7 +2,12 @@
 # Redistribution and use is allowed according to the terms of the BSD license.
 # For details see the accompanying COPYING-CMAKE-SCRIPTS file.
 
-import PyKDE4.kdecore
+try:
+    import PyKDE4.kdecore
+except ImportError,RuntimeError:
+    import sip
+    import PyKDE4.kdecore
+    pass
 
 if __name__ == '__main__':
     try:
--- a/cmake/modules/FindPyQt.py.orig	2017-11-05 02:51:22.000000000 +0100
+++ b/cmake/modules/FindPyQt.py	2019-10-05 18:51:12.803439266 +0200
@@ -3,7 +3,12 @@
 # Redistribution and use is allowed according to the terms of the BSD license.
 # For details see the accompanying COPYING-CMAKE-SCRIPTS file.
 
-import PyQt4.QtCore
+try:
+    import PyQt4.QtCore
+except ImportError,RuntimeError:
+    import sip
+    import PyQt4.QtCore
+    pass
 import os
 import sys
 
@@ -41,8 +46,8 @@
         sip_dir = get_default_sip_dir()
         sip_flags = PyQt4.QtCore.PYQT_CONFIGURATION['sip_flags']
 
-    print('pyqt_version:%06.x' % PyQt4.QtCore.PYQT_VERSION)
-    print('pyqt_version_str:%s' % PyQt4.QtCore.PYQT_VERSION_STR)
+#    print('pyqt_version:%06.x' % PyQt4.QtCore.PYQT_VERSION)
+#    print('pyqt_version_str:%s' % PyQt4.QtCore.PYQT_VERSION_STR)
     print('pyqt_version_tag:%s' % get_qt4_tag(sip_flags))
     print('pyqt_sip_dir:%s' % sip_dir)
     print('pyqt_sip_flags:%s' % sip_flags)
