Index: kfile/kfilewidget.cpp
===================================================================
--- kfile/kfilewidget.cpp	(.../tags/KDE/4.6.0/kdelibs)	(wersja 1219001)
+++ kfile/kfilewidget.cpp	(.../branches/KDE/4.6/kdelibs)	(wersja 1219001)
@@ -54,6 +54,7 @@
 #include <krecentdirs.h>
 #include <kdebug.h>
 #include <kio/kfileitemdelegate.h>
+#include <kde_file.h>
 
 #include <QtGui/QCheckBox>
 #include <QtGui/QDockWidget>
@@ -2294,15 +2295,26 @@
 //     kDebug(kfile_area);
 
     if ((operationMode == KFileWidget::Saving) && (ops->mode() & KFile::File) ) {
-        const QString urlStr = locationEditCurrentText();
+        QString urlStr = locationEditCurrentText();
         if (urlStr.isEmpty())
             return;
 
-        KMimeType::Ptr mime = KMimeType::findByPath(urlStr, 0, true);
-        if (mime && mime->name() != KMimeType::defaultMimeType()) {
-            if (filterWidget->currentFilter() != mime->name() &&
-                filterWidget->filters().indexOf(mime->name()) != -1)
-                filterWidget->setCurrentFilter(mime->name());
+        if( filterWidget->isMimeFilter()) {
+            KMimeType::Ptr mime = KMimeType::findByPath(urlStr, 0, true);
+            if (mime && mime->name() != KMimeType::defaultMimeType()) {
+                if (filterWidget->currentFilter() != mime->name() &&
+                    filterWidget->filters().indexOf(mime->name()) != -1)
+                    filterWidget->setCurrentFilter(mime->name());
+            }
+        } else {
+            QString filename = urlStr.mid( urlStr.lastIndexOf( KDIR_SEPARATOR ) + 1 ); // only filename
+            foreach( const QString& filter, filterWidget->filters()) {
+                QString f = filter.left( filter.indexOf( '|' )); // '*.foo|Foo type' -> '*.foo'
+                if( KMimeType::matchFileName( filename, f )) {
+                    filterWidget->setCurrentFilter( filter );
+                    break;
+                }
+            }
         }
     }
 }
Index: kfile/kfilefiltercombo.h
===================================================================
--- kfile/kfilefiltercombo.h	(.../tags/KDE/4.6.0/kdelibs)	(wersja 1219001)
+++ kfile/kfilefiltercombo.h	(.../branches/KDE/4.6/kdelibs)	(wersja 1219001)
@@ -93,6 +93,12 @@
      * @return all filters (this can be a list of patterns or a list of mimetypes)
      */
     QStringList filters() const;
+    
+    /**
+     * Returns true if the filter has been set using setMimeFilter().
+     * @since 4.6.1
+     */
+    bool isMimeFilter() const;
 
 protected:
     virtual bool eventFilter( QObject*, QEvent* );
Index: kfile/kfilefiltercombo.cpp
===================================================================
--- kfile/kfilefiltercombo.cpp	(.../tags/KDE/4.6.0/kdelibs)	(wersja 1219001)
+++ kfile/kfilefiltercombo.cpp	(.../branches/KDE/4.6/kdelibs)	(wersja 1219001)
@@ -214,4 +214,9 @@
     return d->defaultFilter;
 }
 
+bool KFileFilterCombo::isMimeFilter() const
+{
+    return d->isMimeFilter;
+}
+
 #include "kfilefiltercombo.moc"
Index: kioslave/http/http.cpp
===================================================================
--- kioslave/http/http.cpp	(.../tags/KDE/4.6.0/kdelibs)	(wersja 1219001)
+++ kioslave/http/http.cpp	(.../branches/KDE/4.6/kdelibs)	(wersja 1219001)
@@ -334,7 +334,7 @@
     , m_iSize(NO_SIZE)
     , m_isBusy(false)
     , m_maxCacheAge(DEFAULT_MAX_CACHE_AGE)
-    , m_maxCacheSize(DEFAULT_MAX_CACHE_SIZE/2)
+    , m_maxCacheSize(DEFAULT_MAX_CACHE_SIZE)
     , m_protocol(protocol)
     , m_wwwAuth(0)
     , m_proxyAuth(0)
@@ -3756,7 +3756,7 @@
         if (!cacheFileOpenWrite()) {
             kDebug(7113) << "Error creating cache entry for " << m_request.url.url()<<"!\n";
         }
-        m_maxCacheSize = config()->readEntry("MaxCacheSize", DEFAULT_MAX_CACHE_SIZE) / 2;
+        m_maxCacheSize = config()->readEntry("MaxCacheSize", DEFAULT_MAX_CACHE_SIZE);
     } else if (m_request.responseCode == 304 && m_request.cacheTag.file) {
         if (!mayCache) {
             kDebug(7113) << "This webserver is confused about the cacheability of the data it sends.";
@@ -4928,13 +4928,24 @@
     if (!m_request.cacheTag.file) {
         return;
     }
+
+    // If the file being downloaded is so big that it exceeds the max cache size,
+    // do not cache it! See BR# 244215. NOTE: this can be improved upon in the
+    // future...
+    if (m_iSize >= (m_maxCacheSize * 1024)) {
+        kDebug(7113) << "Caching diabled because content size is too big.";
+        cacheFileClose();
+        return;
+    }
+
     Q_ASSERT(m_request.cacheTag.ioMode == WriteToCache);
     Q_ASSERT(m_request.cacheTag.file->openMode() & QIODevice::WriteOnly);
+
     if (d.isEmpty()) {
         cacheFileClose();
     }
 
-    //### abort if file grows too big! (early abort if we know the size, implement!)
+    //TODO: abort if file grows too big!
 
     // write the variable length text header as soon as we start writing to the file
     if (!m_request.cacheTag.bytesCached) {
Index: kio/kio/accessmanagerreply_p.cpp
===================================================================
--- kio/kio/accessmanagerreply_p.cpp	(.../tags/KDE/4.6.0/kdelibs)	(wersja 1219001)
+++ kio/kio/accessmanagerreply_p.cpp	(.../branches/KDE/4.6/kdelibs)	(wersja 1219001)
@@ -293,11 +293,12 @@
     const QUrl redirectUrl = attribute(QNetworkRequest::RedirectionTargetAttribute).toUrl();
     if (redirectUrl.isValid())
         readHttpResponseHeaders(m_kioJob);
+    else {
+        setAttribute(static_cast<QNetworkRequest::Attribute>(KIO::AccessManager::KioError), errcode);
+        if (errcode)
+            emit error(error());
+    }
 
-    setAttribute(static_cast<QNetworkRequest::Attribute>(KIO::AccessManager::KioError), errcode);
-    if (errcode)
-        emit error(error());
-    
     emit finished();
 }
 
Index: kio/kio/kfileitemdelegate.cpp
===================================================================
--- kio/kio/kfileitemdelegate.cpp	(.../tags/KDE/4.6.0/kdelibs)	(wersja 1219001)
+++ kio/kio/kfileitemdelegate.cpp	(.../branches/KDE/4.6/kdelibs)	(wersja 1219001)
@@ -1666,9 +1666,15 @@
 
     case QEvent::FocusOut:
     {
-        emit commitData(editor);
-        emit closeEditor(editor, NoHint);
-        return true;
+        const QWidget *w = QApplication::activePopupWidget();
+        if (!w || w->parent() != editor)
+        {
+            emit commitData(editor);
+            emit closeEditor(editor, NoHint);
+            return true;
+        }
+        else
+            return false;
     }
 
     default:
Index: kio/kssl/kcm/kcm_ssl.desktop
===================================================================
--- kio/kssl/kcm/kcm_ssl.desktop	(.../tags/KDE/4.6.0/kdelibs)	(wersja 1219001)
+++ kio/kssl/kcm/kcm_ssl.desktop	(.../branches/KDE/4.6/kdelibs)	(wersja 1219001)
@@ -12,7 +12,7 @@
 Name[bg]=Настройки на SSL
 Name[ca]=Preferències SSL
 Name[ca@valencia]=Preferències SSL
-Name[cs]=Nastavení SSLneí
+Name[cs]=Nastavení SSL
 Name[da]=SSL-indstillinger
 Name[de]=SSL-Einstellungen
 Name[el]=Προτιμήσεις SSL
Index: kdecore/services/kmimetype.cpp
===================================================================
--- kdecore/services/kmimetype.cpp	(.../tags/KDE/4.6.0/kdelibs)	(wersja 1219001)
+++ kdecore/services/kmimetype.cpp	(.../branches/KDE/4.6/kdelibs)	(wersja 1219001)
@@ -26,6 +26,7 @@
 #include <kde_file.h> // KDE::stat
 #include <kdeversion.h> // KDE_MAKE_VERSION
 #include <klocale.h>
+#include <kmimetyperepository_p.h>
 #include <kprotocolinfo.h>
 #include <kprotocolinfofactory.h>
 #include <kstandarddirs.h>
@@ -736,6 +737,11 @@
     return QString();
 }
 
+bool KMimeType::matchFileName( const QString &filename, const QString &pattern )
+{
+    return KMimeTypeRepository::matchFileName( filename, pattern );
+}
+
 int KMimeTypePrivate::serviceOffersOffset() const
 {
     return KMimeTypeFactory::self()->serviceOffersOffset(name());
Index: kdecore/services/kmimetyperepository_p.h
===================================================================
--- kdecore/services/kmimetyperepository_p.h	(.../tags/KDE/4.6.0/kdelibs)	(wersja 1219001)
+++ kdecore/services/kmimetyperepository_p.h	(.../branches/KDE/4.6/kdelibs)	(wersja 1219001)
@@ -86,6 +86,11 @@
 
     int sharedMimeInfoVersion();
 
+    /**
+     * @internal
+     */
+    static bool matchFileName( const QString &filename, const QString &pattern );
+
 private: // only for KMimeType and unittests
     friend class KMimeType;
     friend class KMimeFileParserTest;
@@ -128,11 +133,6 @@
      */
     QList<KMimeMagicRule> parseMagicFile(QIODevice* file, const QString& fileName) const;
 
-    /**
-     * @internal (public for unit tests only)
-     */
-    static bool matchFileName( const QString &filename, const QString &pattern );
-
     // Read magic files
     void parseMagic();
 
Index: kdecore/services/kmimetype.h
===================================================================
--- kdecore/services/kmimetype.h	(.../tags/KDE/4.6.0/kdelibs)	(wersja 1219001)
+++ kdecore/services/kmimetype.h	(.../branches/KDE/4.6/kdelibs)	(wersja 1219001)
@@ -400,6 +400,12 @@
     static QString extractKnownExtension( const QString &fileName );
 
     /**
+     * Returns true if the given filename matches the given pattern.
+     * @since 4.6.1
+     */
+    static bool matchFileName( const QString &filename, const QString &pattern );
+
+    /**
      * Returns the version of the installed update-mime-database program
      * (from freedesktop.org shared-mime-info). This is used by unit tests
      * and by the code that writes out icon definitions.
Index: kdecore/localization/all_languages.desktop
===================================================================
--- kdecore/localization/all_languages.desktop	(.../tags/KDE/4.6.0/kdelibs)	(wersja 1219001)
+++ kdecore/localization/all_languages.desktop	(.../branches/KDE/4.6/kdelibs)	(wersja 1219001)
@@ -3173,7 +3173,7 @@
 Name[fi]=Dzongkha
 Name[fr]=Dzongkha
 Name[fy]=Dzongkha
-Name[ga]=Dzongkha
+Name[ga]=Seoinicis
 Name[gl]=Dzongca
 Name[gu]=ઝોન્ખા
 Name[he]=ג'ונקה
Index: plasma/widgets/toolbutton.cpp
===================================================================
--- plasma/widgets/toolbutton.cpp	(.../tags/KDE/4.6.0/kdelibs)	(wersja 1219001)
+++ plasma/widgets/toolbutton.cpp	(.../branches/KDE/4.6/kdelibs)	(wersja 1219001)
@@ -330,7 +330,7 @@
                        const QStyleOptionGraphicsItem *option,
                        QWidget *widget)
 {
-    if (!styleSheet().isNull()) {
+    if (!styleSheet().isNull() || Theme::defaultTheme()->useNativeWidgetStyle()) {
         QGraphicsProxyWidget::paint(painter, option, widget);
         return;
     }
Index: plasma/widgets/scrollwidget.cpp
===================================================================
--- plasma/widgets/scrollwidget.cpp	(.../tags/KDE/4.6.0/kdelibs)	(wersja 1219001)
+++ plasma/widgets/scrollwidget.cpp	(.../branches/KDE/4.6/kdelibs)	(wersja 1219001)
@@ -1422,7 +1422,13 @@
         d->stopAnimations();
         d->adjustScrollbarsTimer->start(200);
         updateGeometry();
-        ensureItemVisible(d->widget.data());
+        if (d->widget.data()->size().width() < viewportGeometry().width() ||
+            d->widget.data()->size().height() < viewportGeometry().height()) {
+            d->widget.data()->setPos(d->minXExtent(),
+                                     d->minYExtent());
+        } else {
+            ensureItemVisible(d->widget.data());
+        }
     } else if (watched == d->widget.data() && event->type() == QEvent::GraphicsSceneMove) {
         d->horizontalScrollBar->blockSignals(true);
         d->verticalScrollBar->blockSignals(true);
Index: plasma/private/focusindicator.cpp
===================================================================
--- plasma/private/focusindicator.cpp	(.../tags/KDE/4.6.0/kdelibs)	(wersja 1219001)
+++ plasma/private/focusindicator.cpp	(.../branches/KDE/4.6/kdelibs)	(wersja 1219001)
@@ -53,6 +53,7 @@
 
 void FocusIndicator::init(QGraphicsWidget *parent)
 {
+    setVisible(!Theme::defaultTheme()->useNativeWidgetStyle());
     setFlag(QGraphicsItem::ItemStacksBehindParent);
     setAcceptsHoverEvents(true);
 
@@ -121,7 +122,8 @@
 
 bool FocusIndicator::eventFilter(QObject *watched, QEvent *event)
 {
-    if (static_cast<QGraphicsWidget *>(watched) != m_parent || !m_parent) {
+    if (Theme::defaultTheme()->useNativeWidgetStyle() ||
+        static_cast<QGraphicsWidget *>(watched) != m_parent || !m_parent ) {
         return false;
     }
 
@@ -274,6 +276,13 @@
 
 void FocusIndicator::syncGeometry()
 {
+    if (Theme::defaultTheme()->useNativeWidgetStyle()) {
+        hide();
+        return;
+    } else if (!isVisible()) {
+        show();
+    }
+
     QRectF geom;
     if (!m_customGeometry.isEmpty()) {
         geom = m_customGeometry;
Index: kdeui/windowmanagement/kwindowsystem_win.cpp
===================================================================
--- kdeui/windowmanagement/kwindowsystem_win.cpp	(.../tags/KDE/4.6.0/kdelibs)	(wersja 1219001)
+++ kdeui/windowmanagement/kwindowsystem_win.cpp	(.../branches/KDE/4.6/kdelibs)	(wersja 1219001)
@@ -417,7 +417,17 @@
     }
 #endif
     // Puts the window in front and activates it.
-    SetForegroundWindow( win );
+    //to bring a window to the front while the user is active in a different apllication we
+    //have to atach our self to the current active window
+    HWND hwndActiveWin = GetForegroundWindow();
+    int  idActive      = GetWindowThreadProcessId(hwndActiveWin, NULL);
+    if ( AttachThreadInput(GetCurrentThreadId(), idActive, TRUE) )
+    {
+        SetForegroundWindow( win );
+        SetFocus( win ); 
+        AttachThreadInput(GetCurrentThreadId(), idActive, FALSE);
+    }
+    
 }
 
 void KWindowSystem::demandAttention( WId win, bool set )
@@ -548,7 +558,16 @@
 
 void KWindowSystem::raiseWindow( WId win )
 {
-    SetWindowPos( win, HWND_TOP, 0, 0, 0, 0, SWP_NOACTIVATE | SWP_NOMOVE | SWP_NOSIZE ); // mhhh?
+
+    //to bring a window to the front while the user is active in a different apllication we
+    //have to atach our self to the current active window
+    HWND hwndActiveWin = GetForegroundWindow();
+    int  idActive      = GetWindowThreadProcessId(hwndActiveWin, NULL);
+    if ( AttachThreadInput(GetCurrentThreadId(), idActive, TRUE) )
+    {
+        SetForegroundWindow( win );
+        AttachThreadInput(GetCurrentThreadId(), idActive, FALSE);
+    }
 }
 
 void KWindowSystem::lowerWindow( WId win )

Zmiany atrybutów dla: .
___________________________________________________________________
Dodane: svn:externals
   + 


