--- kdelibs-4.7.0/kio/kio/kprotocolmanager.cpp~	2011-06-23 16:49:06.000000000 +0200
+++ kdelibs-4.7.0/kio/kio/kprotocolmanager.cpp	2011-07-23 20:34:53.446320740 +0200
@@ -644,6 +644,7 @@
     d->useragent += QString::number(KDE::versionMajor());
     d->useragent += QL1C('.');
     d->useragent += QString::number(KDE::versionMinor());
+    d->useragent += QL1S(" PLDLINUX_VERSION");
   }
   else
   {
