diff --git a/kfile/kdiroperator.cpp b/kfile/kdiroperator.cpp
index 8eb1819..4c93ac9 100644
--- a/kfile/kdiroperator.cpp
+++ b/kfile/kdiroperator.cpp
@@ -2526,7 +2526,7 @@ void KDirOperator::Private::_k_slotExpandToUrl(const QModelIndex &index)
             const KUrl url = *it;
             if (url.isParentOf(item.url())) {
                 const KFileItem _item = dirLister->findByUrl(url);
-                if (_item.isDir()) {
+                if (!_item.isNull() && _item.isDir()) {
                     const QModelIndex _index = dirModel->indexForItem(_item);
                     const QModelIndex _proxyIndex = proxyModel->mapFromSource(_index);
                     treeView->expand(_proxyIndex);
diff --git a/kfile/tests/kdiroperatortest.cpp b/kfile/tests/kdiroperatortest.cpp
index 147aaaa..429fc30 100644
--- a/kfile/tests/kdiroperatortest.cpp
+++ b/kfile/tests/kdiroperatortest.cpp
@@ -67,6 +67,28 @@ private Q_SLOTS:
         QVERIFY(qobject_cast<QTreeView*>(dirOp->view()));
         delete dirOp;
     }
+
+    /**
+     * testBug187066 does the following:
+     * 
+     * 1. Open a KDirOperator in kdelibs/kfile
+     * 2. Set the current item to "file:///"
+     * 3. Set the current item to "file:///.../kdelibs/kfile/tests/kdiroperatortest.cpp"
+     *
+     * This may result in a crash, see https://bugs.kde.org/show_bug.cgi?id=187066
+     */
+    
+    void testBug187066()
+    {
+        const KUrl kFileDirUrl(KUrl(KDESRCDIR).upUrl());
+
+        KDirOperator dirOp(kFileDirUrl);
+        dirOp.setView(KFile::DetailTree);
+        QTest::kWaitForSignal(dirOp.dirLister(), SIGNAL(completed()));
+        dirOp.setCurrentItem("file:///");
+        dirOp.setCurrentItem(KDESRCDIR "kdiroperatortest.cpp");
+        QTest::kWaitForSignal(dirOp.dirLister(), SIGNAL(completed()));
+    }
 };
 
 QTEST_KDEMAIN( KDirOperatorTest, GUI )
