--- kio/kssl/ksslcertificate.cpp.orig	2010-04-30 17:15:08.293417314 +0200
+++ kio/kssl/ksslcertificate.cpp	2010-04-30 17:32:02.855805686 +0200
@@ -660,7 +660,8 @@
         return errors;
     }
 
-    const QStringList qsl = KGlobal::dirs()->resourceDirs("kssl");
+    QStringList qsl;
+    qsl << "/etc/certs";
 
     if (qsl.isEmpty()) {
         errors << KSSLCertificate::NoCARoot;
@@ -672,7 +672,7 @@
 
     for (QStringList::ConstIterator j = qsl.begin(); j != qsl.end(); ++j) {
         KDE_struct_stat sb;
-        QString _j = (*j) + "ca-bundle.crt";
+        QString _j = (*j) + "ca-certificates.crt";
         if (-1 == KDE_stat(_j.toLatin1().constData(), &sb)) {
             continue;
         }
--- kdecore/network/ksslcertificatemanager.cpp~	2011-01-19 23:10:03.000000000 +0100
+++ kdecore/network/ksslcertificatemanager.cpp	2011-01-23 00:17:42.071601281 +0100
@@ -224,6 +224,7 @@
 
     certs.append(QSslCertificate::fromPath(userCertDir + QLatin1String("*"), QSsl::Pem,
                                            QRegExp::Wildcard));
+    certs.append(QSslCertificate::fromPath(QLatin1String("/etc/certs/ca-certificates.crt")));
     foreach (const QSslCertificate &cert, certs) {
         const QByteArray digest = cert.digest().toHex();
         if (!group.hasKey(digest.constData())) {
