--- cmake/modules/FindBoost.cmake.old	2008-07-23 11:56:52.633942000 +0200
+++ cmake/modules/FindBoost.cmake	2008-07-23 11:57:25.732010500 +0200
@@ -211,12 +211,15 @@
     # cdrom inserted it will popup a very annoying dialog
     #D:/boost/include
     /sw/local/include
+    /usr/include
   )
 
   SET(_boost_LIBRARIES_SEARCH_DIRS
     C:/boost/lib
     "C:/Program Files/boost/boost_${Boost_MINIMUM_VERSION}/lib"
     /sw/local/lib
+    /usr/lib
+    /usr/lib64
   )
 
   IF( NOT $ENV{BOOST_ROOT} STREQUAL "" )
