--- x/solid/solid/backends/upower/upowerdevice.cpp	2013-03-01 07:26:14.555774261 +0100
+++ x/solid/solid/backends/upower/upowerdevice.cpp	2013-03-26 20:13:35.925180209 +0100
@@ -83,7 +83,7 @@
         case Solid::DeviceInterface::GenericInterface:
             return true;
         case Solid::DeviceInterface::Battery:
-            return (uptype == 2 || uptype == 3);
+            return (uptype == 2 || uptype == 3 || uptype == 5 || uptype == 6);
         case Solid::DeviceInterface::AcAdapter:
             return (uptype == 1);
         default:
