Index: khtml/khtml_part.cpp
===================================================================
--- khtml/khtml_part.cpp	(.../tags/KDE/4.2.1/kdelibs)	(wersja 935993)
+++ khtml/khtml_part.cpp	(.../branches/KDE/4.2/kdelibs)	(wersja 935993)
@@ -90,6 +90,7 @@
 #include <klocale.h>
 #include <kmessagebox.h>
 #include <kstandardaction.h>
+#include <kstandardguiitem.h>
 #include <kactioncollection.h>
 #include <kfiledialog.h>
 #include <kmimetypetrader.h>
@@ -415,14 +416,18 @@
           // Nobody else does it...
           d->m_paIncZoomFactor->setShortcut( KShortcut("CTRL++; CTRL+=") );
           d->m_paDecZoomFactor->setShortcut( QKeySequence(Qt::CTRL + Qt::Key_Minus) );
+          d->m_paIncZoomFactor->setShortcutContext( Qt::WidgetWithChildrenShortcut );
+          d->m_paDecZoomFactor->setShortcutContext( Qt::WidgetWithChildrenShortcut );
       }
   }
 
   d->m_paFind = actionCollection()->addAction( KStandardAction::Find, "find", this, SLOT( slotFind() ) );
+  d->m_paFind->setShortcutContext( Qt::WidgetWithChildrenShortcut ); // default context conflicts when splitting konqueror
   d->m_paFind->setWhatsThis( i18n( "Find text<br /><br />"
                                    "Shows a dialog that allows you to find text on the displayed page." ) );
 
   d->m_paFindNext = actionCollection()->addAction( KStandardAction::FindNext, "findNext", this, SLOT( slotFindNext() ) );
+  d->m_paFindNext->setShortcutContext( Qt::WidgetWithChildrenShortcut );
   d->m_paFindNext->setWhatsThis( i18n( "Find next<br /><br />"
                                        "Find the next occurrence of the text that you "
                                        "have found using the <b>Find Text</b> function" ) );
@@ -477,6 +482,7 @@
   d->m_paToggleCaretMode = new KToggleAction(i18n("Toggle Caret Mode"), this );
   actionCollection()->addAction( "caretMode", d->m_paToggleCaretMode );
   d->m_paToggleCaretMode->setShortcut( QKeySequence(Qt::Key_F7) );
+  d->m_paToggleCaretMode->setShortcutContext( Qt::WidgetWithChildrenShortcut );
   connect( d->m_paToggleCaretMode, SIGNAL( triggered( bool ) ), this, SLOT(slotToggleCaretMode()) );
   d->m_paToggleCaretMode->setChecked(isCaretMode());
   if (parentPart())
@@ -6602,10 +6608,14 @@
     int response = KMessageBox::Cancel;
     if (!message.isEmpty())
     {
+            // Dangerous flag makes the Cancel button the default
             response = KMessageBox::warningContinueCancel( 0,
                                                            message.subs(Qt::escape(linkURL.prettyUrl())).toString(),
                                                            i18n( "Security Warning" ),
-                                                           KGuiItem(button));
+                                                           KGuiItem(button),
+                                                           KStandardGuiItem::cancel(),
+                                                           QString(), // no don't ask again info
+                                                           KMessageBox::Notify | KMessageBox::Dangerous );
     }
     else
     {
Index: khtml/html/html_formimpl.cpp
===================================================================
--- khtml/html/html_formimpl.cpp	(.../tags/KDE/4.2.1/kdelibs)	(wersja 935993)
+++ khtml/html/html_formimpl.cpp	(.../branches/KDE/4.2/kdelibs)	(wersja 935993)
@@ -2314,6 +2314,12 @@
         setRecalcListItems();
 }
 
+void HTMLSelectElementImpl::removeChildren()
+{
+    HTMLGenericFormElementImpl::removeChildren();
+    setRecalcListItems();
+}
+
 NodeImpl *HTMLSelectElementImpl::appendChild ( NodeImpl *newChild, int &exceptioncode )
 {
     NodeImpl* const result = HTMLGenericFormElementImpl::appendChild(newChild, exceptioncode);
Index: khtml/html/html_tableimpl.h
===================================================================
--- khtml/html/html_tableimpl.h	(.../tags/KDE/4.2.1/kdelibs)	(wersja 935993)
+++ khtml/html/html_tableimpl.h	(.../branches/KDE/4.2/kdelibs)	(wersja 935993)
@@ -216,6 +216,10 @@
         //else things are unchanged.
     }
 
+    void clear() {
+        ptr = 0;
+    }
+
     void operator =(ChildType* child) {
         ptr = child;
     }
@@ -277,6 +281,7 @@
     virtual NodeImpl *insertBefore ( NodeImpl *newChild, NodeImpl *refChild, int &exceptioncode );
     virtual void      replaceChild ( NodeImpl *newChild, NodeImpl *oldChild, int &exceptioncode );
     virtual void      removeChild ( NodeImpl *oldChild, int &exceptioncode );
+    virtual void      removeChildren();
     virtual NodeImpl *appendChild ( NodeImpl *newChild, int &exceptioncode );
     
     virtual void parseAttribute(AttributeImpl *attr);
Index: khtml/html/html_formimpl.h
===================================================================
--- khtml/html/html_formimpl.h	(.../tags/KDE/4.2.1/kdelibs)	(wersja 935993)
+++ khtml/html/html_formimpl.h	(.../branches/KDE/4.2/kdelibs)	(wersja 935993)
@@ -419,6 +419,7 @@
     virtual NodeImpl *insertBefore ( NodeImpl *newChild, NodeImpl *refChild, int &exceptioncode );
     virtual void      replaceChild ( NodeImpl *newChild, NodeImpl *oldChild, int &exceptioncode );
     virtual void      removeChild ( NodeImpl *oldChild, int &exceptioncode );
+    virtual void      removeChildren();
     virtual NodeImpl *appendChild ( NodeImpl *newChild, int &exceptioncode );
     virtual NodeImpl *addChild( NodeImpl* newChild );
 
Index: khtml/html/html_tableimpl.cpp
===================================================================
--- khtml/html/html_tableimpl.cpp	(.../tags/KDE/4.2.1/kdelibs)	(wersja 935993)
+++ khtml/html/html_tableimpl.cpp	(.../branches/KDE/4.2/kdelibs)	(wersja 935993)
@@ -395,7 +395,16 @@
     handleChildRemove( oldChild );
     HTMLElementImpl::removeChild( oldChild, exceptioncode);
 }
- 
+
+void HTMLTableElementImpl::removeChildren()
+{
+    HTMLElementImpl::removeChildren();
+    tCaption.clear();
+    head.clear();
+    foot.clear();
+    firstBody.clear();
+}
+
 static inline bool isTableCellAncestor(NodeImpl* n)
 {
     return n->id() == ID_THEAD || n->id() == ID_TBODY ||
Index: khtml/xml/dom_nodeimpl.h
===================================================================
--- khtml/xml/dom_nodeimpl.h	(.../tags/KDE/4.2.1/kdelibs)	(wersja 935993)
+++ khtml/xml/dom_nodeimpl.h	(.../branches/KDE/4.2/kdelibs)	(wersja 935993)
@@ -545,7 +545,7 @@
     virtual bool hasChildNodes (  ) const;
 
     // Other methods (not part of DOM)
-    void removeChildren();
+    virtual void removeChildren();
     void cloneChildNodes(NodeImpl *clone);
 
     virtual void setFirstChild(NodeImpl *child);
Index: khtml/xml/dom_docimpl.cpp
===================================================================
--- khtml/xml/dom_docimpl.cpp	(.../tags/KDE/4.2.1/kdelibs)	(wersja 935993)
+++ khtml/xml/dom_docimpl.cpp	(.../branches/KDE/4.2/kdelibs)	(wersja 935993)
@@ -2828,8 +2828,8 @@
 
     m_imageLoadEventDispatchingList = m_imageLoadEventDispatchSoonList;
     m_imageLoadEventDispatchSoonList.clear();
-    for (QLinkedListIterator<HTMLImageElementImpl*> it(m_imageLoadEventDispatchingList); it.hasNext(); )
-        it.next()->dispatchLoadEvent();
+    while (!m_imageLoadEventDispatchingList.isEmpty())
+        m_imageLoadEventDispatchingList.takeFirst()->dispatchLoadEvent();
     m_imageLoadEventDispatchingList.clear();
 }
 
Index: kate/plugins/insertfile/ktexteditor_insertfile.desktop
===================================================================
--- kate/plugins/insertfile/ktexteditor_insertfile.desktop	(.../tags/KDE/4.2.1/kdelibs)	(wersja 935993)
+++ kate/plugins/insertfile/ktexteditor_insertfile.desktop	(.../branches/KDE/4.2/kdelibs)	(wersja 935993)
@@ -42,7 +42,7 @@
 Name[gu]=ફાઇલ ઉમેરો
 Name[he]=הוספת קובץ
 Name[hi]=फ़ाइल प्रविष्ट करें
-Name[hne]=फाइल घुसाव
+Name[hne]=फाइल भरव
 Name[hsb]=Dokument zasunyć
 Name[hu]=Fájl beszúrása
 Name[is]=Skjóta inn skrá
@@ -120,7 +120,7 @@
 Comment[gu]=કર્સર સ્થિતિ પર કોઇપણ વાંચી શકાય તેવી ફાઇલ દાખલ કરો
 Comment[he]=מוסיף כל קובץ הניתן לקריאה במיקום הסמן
 Comment[hi]=संकेतक के स्थान पर कोई भी पढ़ने योग्य फ़ाइल प्रविष्ट करें
-Comment[hne]=संकेतक के जगह फेर कोनो पढ़े जा सके फाइल घुसाव
+Comment[hne]=संकेतक के जगह तिर कोनो पढ़े जा सके फाइल भरव
 Comment[hr]=Na položaju pokazivača umetnite bilo koju čitljivu datoteku
 Comment[hsb]=Zasunje lubowólnu čitajomnu dataju na poziciji cursora
 Comment[hu]=Tetszőleges olvasható fájl beszúrása a kurzorpozíciónál
Index: kate/plugins/wordcompletion/ktexteditor_docwordcompletion.desktop
===================================================================
--- kate/plugins/wordcompletion/ktexteditor_docwordcompletion.desktop	(.../tags/KDE/4.2.1/kdelibs)	(wersja 935993)
+++ kate/plugins/wordcompletion/ktexteditor_docwordcompletion.desktop	(.../branches/KDE/4.2/kdelibs)	(wersja 935993)
@@ -150,6 +150,7 @@
 Comment[pt_BR]=Complementação direcional ou baseada em popup, a partir de palavras do documento
 Comment[ro]=Propune completarea cuvintelor din document dintr-o listă popup sau direcțională
 Comment[ru]=Автозавершение слов в документе
+Comment[se]=Ollášuhttin mii oidno čállingiettis vai listtus mii bahccá
 Comment[sk]=Dopĺňanie slov v dokumente priame alebo pomocou dialógu
 Comment[sl]=Neposredno ali pojavno dopolnjevanje iz besed v dokumentu
 Comment[sr]=Дирекционо или искачуће допуњавање према речима из документа
Index: kate/plugins/timedate/ktexteditor_timedate.desktop
===================================================================
--- kate/plugins/timedate/ktexteditor_timedate.desktop	(.../tags/KDE/4.2.1/kdelibs)	(wersja 935993)
+++ kate/plugins/timedate/ktexteditor_timedate.desktop	(.../branches/KDE/4.2/kdelibs)	(wersja 935993)
@@ -117,7 +117,7 @@
 Comment[gu]=હાલનો સમય & તારીખ ઉમેરો
 Comment[he]=הוספת התאריך והשעה הנוכחיים
 Comment[hi]=वर्तमान समय व तिथि प्रविष्ट करें
-Comment[hne]=अभी हाल के समय व तारीक घुसाव
+Comment[hne]=अभी हाल के समय व तारीक भरव
 Comment[hsb]=Zasunje aktualny čas a datum
 Comment[hu]=Az aktuális dátum és idő beszúrása
 Comment[hy]=Մտցրեք ընթացիք Ժամը և Ամսաթիվը
Index: kate/syntax/data/lilypond.xml
===================================================================
--- kate/syntax/data/lilypond.xml	(.../tags/KDE/4.2.1/kdelibs)	(wersja 935993)
+++ kate/syntax/data/lilypond.xml	(.../branches/KDE/4.2/kdelibs)	(wersja 935993)
@@ -14,7 +14,7 @@
   <!ENTITY markupwithtextargs "markup|bold|(rounded-)?box|bracket|caps|(center|general|left|right)-align|circle|((center|dir|left|right)-)?column|combine|concat|dynamic|fill-line|finger|fontCaps|(abs-)?fontsize|fraction|halign|hbracket|hcenter-in|hcenter|hspace|huge|italic|justify|larger?|line|lower|magnify|medium|normal-size-(sub|super)|normal-text|normalsize|number|on-the-fly|override|pad-(around|markup|to-box|x)|page-ref|postscript|put-adjacent|raise|roman|rotate|sans|small(er)?|smallCaps|sub|super|teeny|text|tiny|translate(-scaled)?|transparent|typewriter|underline|upright|vcenter|whiteout|with-(color|dimensions|url)|wordwrap|(markup|column-|justified-|override-|wordwrap-)lines|wordwrap-(string-)?internal">
   <!ENTITY deprecatedmarkup "bigger|h?center">
   <!ENTITY headervars "dedication|(sub){,2}title|poet|composer|meter|opus|arranger|instrument|piece|breakbefore|copyright|tagline|mutopia(title|composer|poet|opus|instrument)|date|enteredby|source|style|maintainer(Email|Web)?|moreInfo|lastupdated|texidoc|footer">
-  <!ENTITY papervars "annotate-spacing|(print-)?first-page-number|print-page-number|paper-(width|height)|(top|bottom|left|right)-margin|line-width|(head|foot)-separation|page-top-space|ragged-(bottom|last(-bottom)?|right)|page-count|between-system-(space|padding)|page-breaking-between-system-padding|horizontal-shift|(before|after|between)-title-space|print-all-headers|indent|force-assignment|input-encoding|output-scale|blank(-after-score|-last)?-page-force|page-limit-inter-system-space(-factor)?|(systemSeparator|(even|odd)(Footer|Header)|(book|score|toc)Title|tocItem)Markup">
+  <!ENTITY papervars "annotate-spacing|(print-)?first-page-number|print-page-number|paper-(width|height)|(top|bottom|left|right)-margin|line-width|(head|foot)-separation|page-top-space|ragged-(bottom|last(-bottom)?|right)|page-count|between-system-(space|padding)|page-breaking-between-system-padding|horizontal-shift|(before|after|between)-title-space|print-all-headers|indent|force-assignment|input-encoding|output-scale|blank(-after-score|-last)?-page-force|page-limit-inter-system-space(-factor)?|system-separator-markup|((even|odd)(Footer|Header)|(book|score|toc)Title|tocItem)Markup">
   <!ENTITY layoutvars "system-count|(short-)?indent">
   <!ENTITY toplevelvars "dash(Hat|Plus|Dash|Bar|Larger|Dot|Underscore)|fermataMarkup|pipeSymbol|slashSeparator">
   <!ENTITY performer "Beam|Control_track|Drum_note|Dynamic|Key|Lyric|Note|Piano_pedal|Slur|Staff|Swallow|Tempo|Tie|Time_signature">
Index: kate/syntax/data/ruby.xml
===================================================================
--- kate/syntax/data/ruby.xml	(.../tags/KDE/4.2.1/kdelibs)	(wersja 935993)
+++ kate/syntax/data/ruby.xml	(.../branches/KDE/4.2/kdelibs)	(wersja 935993)
@@ -327,8 +327,8 @@
 			
 			<!-- Internal context used by check_div_2 -->
 			<context name="check_div_2_internal" attribute="Normal Text" fallthrough="true" fallthroughContext="#pop#pop" lineEndContext="#pop#pop">
-				<DetectChar attribute="Operator" char="%" context="#pop#pop"/>
-				<RegExpr attribute="Operator" String="/(?=\s)" context="#pop#pop"/>
+<!-- 				<DetectChar attribute="Operator" char="%" context="#pop#pop"/> -->
+				<RegExpr attribute="Operator" String="[/%](?=\s)" context="#pop#pop"/>
 			</context>
 			
 			<!-- Same as check_div_2, but with double pop to exit the surrounding context -->
@@ -359,12 +359,10 @@
 				<RegExpr attribute="String" String="\\\&quot;" context="#stay"/>
 				<RegExpr attribute="Substitution" String="#@{1,2}" context="Short Subst"/>
 				<Detect2Chars attribute="Substitution" char="#" char1="{" context="Subst"/>
-				<!--HlCChar attribute="Char" context="#pop"/-->
 				<DetectChar char="&quot;" attribute="String" context="check_div_1_pop"/>
 			</context>
 			
 			<context name="Apostrophed String" attribute="Raw String" lineEndContext="#stay">
-				<!-- <HlCChar attribute="Char" context="#pop"/> -->
 				<StringDetect attribute="String" String="\\" context="#stay"/>
 				<RegExpr attribute="String" String="\\\'" context="#stay"/>
 				<DetectChar char="'" attribute="Raw String" context="check_div_1_pop"/>
@@ -375,7 +373,6 @@
 				<RegExpr attribute="String" String="\\\`" context="#stay"/>
 				<RegExpr attribute="Substitution" String="#@{1,2}" context="Short Subst"/>
 				<Detect2Chars attribute="Substitution" char="#" char1="{" context="Subst"/>
-				<HlCChar attribute="Char" context="check_div_1_pop"/>
 				<DetectChar char="`" attribute="Command" context="check_div_1_pop"/>
 			</context>
 			
Index: kate/tests/highlight.rb
===================================================================
--- kate/tests/highlight.rb	(.../tags/KDE/4.2.1/kdelibs)	(wersja 935993)
+++ kate/tests/highlight.rb	(.../branches/KDE/4.2/kdelibs)	(wersja 935993)
@@ -422,10 +422,17 @@
 10 % 4          # modulo
 
 foo%4           # modulo
-foo %4          # modulo
+# foo %4          # illegal %string
 foo% 4          # modulo
 foo % 4         # modulo
 
+foo % (4)       # modulo
+
+foo %(4)        # %string
+foo %q(4)       # %string
+foo %Q(4)       # %string
+foo %%4%        # %string
+
 foo = %|blah|   # GDL input
 foo % %|blah|   # modulo and GDL
 
@@ -470,6 +477,10 @@
 %x{ls -l |
 grep test }
 
+# alternative syntax
+`ls -l`
+`echo ' '`
+
 # token array
 %w{one two three}
 %w(one two three)
Index: kate/completion/katecompletionmodel.cpp
===================================================================
--- kate/completion/katecompletionmodel.cpp	(.../tags/KDE/4.2.1/kdelibs)	(wersja 935993)
+++ kate/completion/katecompletionmodel.cpp	(.../branches/KDE/4.2/kdelibs)	(wersja 935993)
@@ -573,7 +573,9 @@
 
 void KateCompletionModel::createGroups()
 {
-  clearGroups(false); //Reset is done below
+  //After clearing the model, it has to be reset, else we will be in an invalid state while inserting
+  //new groups.
+  clearGroups(true);
 
   bool has_groups=false;
   foreach (CodeCompletionModel* sourceModel, m_completionModels) {
@@ -945,8 +947,10 @@
     updateBestMatches();
   }
 
+  if(changeType == Broaden)
+    reset(); //There is a bug in changeCompletions() notification code of the "Broaden" operation, so just reset
+
   clearExpanding(); //We need to do this, or be aware of expanding-widgets while filtering.
-// reset();
   emit contentGeometryChanged();
 }
 
@@ -974,7 +978,7 @@
 #define COMPLETE_DELETE \
   if (rowDeleteStart != -1) { \
     if (!g->isEmpty) \
-      deleteRows(g, filtered, index - rowDeleteStart, rowDeleteStart); \
+      deleteRows(g, filtered, index - rowDeleteStart, rowDeleteStart, changeType != Broaden); \
     filterIndex -= index - rowDeleteStart + 1; \
     index -= index - rowDeleteStart; \
     rowDeleteStart = -1; \
@@ -982,7 +986,7 @@
 
 #define COMPLETE_ADD \
   if (rowAddStart != -1) { \
-    addRows(g, filtered, rowAddStart, rowAdd); \
+    addRows(g, filtered, rowAddStart, rowAdd, changeType != Broaden); \
     filterIndex += rowAdd.count(); \
     index += rowAdd.count(); \
     rowAddStart = -1; \
@@ -1116,7 +1120,9 @@
   COMPLETE_DELETE
   COMPLETE_ADD
 
-  hideOrShowGroup(g);
+  //Only notify the model if the changetype is not broaden, else it will barf on it
+  //because the code above has a bug in that case. We do a reset() anyway afterwards in that case.
+  hideOrShowGroup(g, changeType != Broaden);
 }
 
 int KateCompletionModel::Group::orderNumber() const {
@@ -1150,7 +1156,7 @@
     return orderNumber() < other->orderNumber();
 }
 
-void KateCompletionModel::hideOrShowGroup(Group* g)
+void KateCompletionModel::hideOrShowGroup(Group* g, bool notifyModel)
 {
   if( g == m_argumentHints ) {
     emit argumentHintsChanged();
@@ -1164,10 +1170,10 @@
       g->isEmpty = true;
       int row = m_rowTable.indexOf(g);
       if (row != -1) {
-        if (hasGroups())
+        if (hasGroups() && notifyModel)
           beginRemoveRows(QModelIndex(), row, row);
         m_rowTable.removeAt(row);
-        if (hasGroups())
+        if (hasGroups() && notifyModel)
           endRemoveRows();
         m_emptyGroups.append(g);
       } else {
@@ -1190,23 +1196,27 @@
         row = a+1;
       }
 
-      if (hasGroups())
-        beginInsertRows(QModelIndex(), row, row);
-      else
-        beginInsertRows(QModelIndex(), 0, g->filtered.count());
+      if(notifyModel) {
+        if (hasGroups())
+          beginInsertRows(QModelIndex(), row, row);
+        else
+          beginInsertRows(QModelIndex(), 0, g->filtered.count());
+      }
       m_rowTable.insert(row, g);
-      endInsertRows();
+      if(notifyModel)
+        endInsertRows();
       m_emptyGroups.removeAll(g);
     }
   }
 }
 
-void KateCompletionModel::deleteRows( Group* g, QMutableListIterator<Item> & filtered, int countBackwards, int startRow )
+void KateCompletionModel::deleteRows( Group* g, QMutableListIterator<Item> & filtered, int countBackwards, int startRow, bool notify )
 {
   QModelIndex groupIndex = indexForGroup(g);
   Q_ASSERT(!hasGroups() || groupIndex.isValid());
 
-  beginRemoveRows(groupIndex, startRow, startRow + countBackwards - 1);
+  if(notify)
+    beginRemoveRows(groupIndex, startRow, startRow + countBackwards - 1);
 
   for (int i = 0; i < countBackwards; ++i) {
     filtered.remove();
@@ -1221,12 +1231,13 @@
 
     filtered.previous();
   }
-
-  endRemoveRows();
+  
+  if(notify)
+    endRemoveRows();
 }
 
 
-void KateCompletionModel::addRows( Group * g, QMutableListIterator<Item> & filtered, int startRow, const QList<Item> & newItems )
+void KateCompletionModel::addRows( Group * g, QMutableListIterator<Item> & filtered, int startRow, const QList<Item> & newItems, bool notify )
 {
   //bool notify = true;
 
@@ -1237,13 +1248,13 @@
 
   kDebug( 13035 ) << "Group" << g->title << "addRows" << startRow << "to " << (startRow + newItems.count() - 1);
 
-  //if (notify)
+  if (notify)
     beginInsertRows(groupIndex, startRow, startRow + newItems.count() - 1);
 
   for (int i = 0; i < newItems.count(); ++i)
     filtered.insert(newItems[i]);
 
-  //if (notify)
+  if (notify)
     endInsertRows();
 }
 
@@ -1997,6 +2008,7 @@
 
 //Updates the best-matches group
 void KateCompletionModel::updateBestMatches() {
+  int maxMatches = 300; //We cannot do too many operations here, because they are all executed whenever a character is added. Would be nice if we could split the operations up somewhat using a timer.
 
   m_updateBestMatchesTimer->stop();
   //Maps match-qualities to ModelRows paired together with the BestMatchesCount returned by the items.
@@ -2020,6 +2032,9 @@
       }
       
       ++row;
+      --maxMatches;
+      if(maxMatches < 0)
+        break;
     }
     
     if(!rowsForQuality.isEmpty()) {
@@ -2043,7 +2058,6 @@
   }
   
   ///@todo Cache the CodeCompletionModel::BestMatchesCount
-  int maxMatches = 50; //We cannot do too many operations here, because they are all executed whenever a character is added. Would be nice if we could split the operations up somewhat using a timer.
   foreach (Group* g, m_rowTable) {
     if( g == m_bestMatches )
       continue;
Index: kate/completion/katecompletionmodel.h
===================================================================
--- kate/completion/katecompletionmodel.h	(.../tags/KDE/4.2.1/kdelibs)	(wersja 935993)
+++ kate/completion/katecompletionmodel.h	(.../branches/KDE/4.2/kdelibs)	(wersja 935993)
@@ -282,7 +282,7 @@
     QSet<Group*> deleteItems(const QModelIndex& i);
     Group* createItem(const HierarchicalModelHandler&, const QModelIndex& i, bool notifyModel = false);
     void clearGroups(bool reset = true);
-    void hideOrShowGroup(Group* g);
+    void hideOrShowGroup(Group* g, bool notifyModel = true);
     /// When forceGrouping is enabled, all given attributes will be used for grouping, regardless of the completion settings.
     Group* fetchGroup(int attribute, const QString& scope = QString(), bool forceGrouping = false);
     //If this returns nonzero on an index, the index is the header of the returned group
@@ -300,8 +300,8 @@
 
     void changeCompletions(Group* g, changeTypes changeType);
 
-    void deleteRows(Group* g, QMutableListIterator<Item>& filtered, int countBackwards, int startRow);
-    void addRows(Group* g, QMutableListIterator<Item>& filtered, int startRow, const QList<Item>& newItems);
+    void deleteRows(Group* g, QMutableListIterator<Item>& filtered, int countBackwards, int startRow, bool notify);
+    void addRows(Group* g, QMutableListIterator<Item>& filtered, int startRow, const QList<Item>& newItems, bool notify);
 
     bool hasGroups() const;
     bool hasCompletionModel() const;
Index: kate/completion/katecompletionwidget.cpp
===================================================================
--- kate/completion/katecompletionwidget.cpp	(.../tags/KDE/4.2.1/kdelibs)	(wersja 935993)
+++ kate/completion/katecompletionwidget.cpp	(.../branches/KDE/4.2/kdelibs)	(wersja 935993)
@@ -203,6 +203,9 @@
 void KateCompletionWidget::rowsInserted(const QModelIndex& parent, int rowFrom, int rowEnd)
 {
   m_entryList->setAnimated(false);
+  if(!model()->isGroupingEnabled())
+    return;
+
   if (!parent.isValid())
     for (int i = rowFrom; i <= rowEnd; ++i)
       m_entryList->expand(m_presentationModel->index(i, 0, parent));
@@ -450,7 +453,12 @@
               h = localHeight;
           }
           baseHeight += h;
+          if(baseHeight > maxBaseHeight)
+            break;
         }
+
+        if(baseHeight > maxBaseHeight)
+          break;
       }
     }
     
@@ -461,11 +469,18 @@
   
   if(m_entryList->horizontalScrollBar()->isVisible())
     baseHeight += m_entryList->horizontalScrollBar()->height();
-  
+
+
   if(baseHeight < minBaseHeight)
     baseHeight = minBaseHeight;
-  if(baseHeight > maxBaseHeight)
+  if(baseHeight > maxBaseHeight) {
     baseHeight = maxBaseHeight;
+    m_entryList->setVerticalScrollBarPolicy(Qt::ScrollBarAsNeeded);
+  }else{
+    //Somewhere there seems to be a bug that makes QTreeView add a scroll-bar
+    //even if the content exactly fits in. So forcefully disable the scroll-bar in that case
+    m_entryList->setVerticalScrollBarPolicy(Qt::ScrollBarAlwaysOff);
+}
   
   int newExpandingAddedHeight = 0;
   
@@ -506,7 +521,8 @@
   geom.setHeight(finalHeight);
 
   setGeometry(geom);
-    m_entryList->resize(m_entryList->width(), height() - 2*frameWidth());
+
+  m_entryList->resize(m_entryList->width(), finalHeight - 2*frameWidth());
 }
 
 
@@ -517,25 +533,34 @@
 
   KTextEditor::Cursor cursor = view()->cursorPosition();
 
+  QList<KTextEditor::CodeCompletionModel*> checkCompletionRanges = m_completionRanges.keys();
+
   //Check the models and eventuall abort some
-  QMutableMapIterator<KTextEditor::CodeCompletionModel*, KateSmartRange*> i(m_completionRanges);
-  while (i.hasNext()) {
-      i.next();
-      KTextEditor::CodeCompletionModel *model = i.key();
-      KateSmartRange* range = i.value();
+  for(QList<KTextEditor::CodeCompletionModel*>::iterator it = checkCompletionRanges.begin();
+      it != checkCompletionRanges.end(); ++it) {
+      if(!m_completionRanges.contains(*it))
+        continue;
+
+      KTextEditor::CodeCompletionModel *model = *it;
+      KateSmartRange* range = m_completionRanges[*it];
+
       modelController(model)->updateCompletionRange(view(), *range);
       QString currentCompletion = modelController(model)->filterString(view(), *range, view()->cursorPosition());
       bool abort = modelController(model)->shouldAbortCompletion(view(), *range, currentCompletion);
 
+      if(!m_completionRanges.contains(*it))
+        continue;
+
       if (abort) {
         if (m_completionRanges.count() == 1) {
           //last model - abort whole completion
           abortCompletion();
           return;
         } else {
+          delete m_completionRanges[*it];
+          m_completionRanges.remove(*it);
+
           modelController(model)->aborted(view());
-          i.remove();
-          delete range;
           m_presentationModel->removeCompletionModel(model);
         }
       } else {
Index: kate/view/kateviewinternal.cpp
===================================================================
--- kate/view/kateviewinternal.cpp	(.../tags/KDE/4.2.1/kdelibs)	(wersja 935993)
+++ kate/view/kateviewinternal.cpp	(.../branches/KDE/4.2/kdelibs)	(wersja 935993)
@@ -1201,11 +1201,6 @@
 
 void KateViewInternal::end( bool sel )
 {
-  if (m_view->isCompletionActive()) {
-    view()->completionWidget()->bottom();
-    return;
-  }
-
   QMutexLocker lock(m_doc->smartMutex());
 
   KateTextLayout layout = currentLayout();
Index: kfile/kfilewidget.cpp
===================================================================
--- kfile/kfilewidget.cpp	(.../tags/KDE/4.2.1/kdelibs)	(wersja 935993)
+++ kfile/kfilewidget.cpp	(.../branches/KDE/4.2/kdelibs)	(wersja 935993)
@@ -836,12 +836,13 @@
     } else if (locationEditCurrentTextList.count()) {
         // if we are on file or files mode, and we have an absolute url written by
         // the user, convert it to relative
-        KUrl _url(locationEditCurrentText);
         if (!locationEditCurrentText.isEmpty() && !(mode & KFile::Directory) &&
-            (QDir::isAbsolutePath(locationEditCurrentText) || !_url.protocol().isEmpty())) {
+            (QDir::isAbsolutePath(locationEditCurrentText) ||
+             containsProtocolSection(locationEditCurrentText))) {
+
             QString fileName;
-            KUrl url(_url);
-            KIO::StatJob *statJob = KIO::stat(_url, KIO::HideProgressInfo);
+            KUrl url(locationEditCurrentText);
+            KIO::StatJob *statJob = KIO::stat(url, KIO::HideProgressInfo);
             bool res = KIO::NetAccess::synchronousRun(statJob, 0);
             if (res) {
                 if (!statJob->statResult().isDir()) {
Index: kioslave/file/file.cpp
===================================================================
--- kioslave/file/file.cpp	(.../tags/KDE/4.2.1/kdelibs)	(wersja 935993)
+++ kioslave/file/file.cpp	(.../branches/KDE/4.2/kdelibs)	(wersja 935993)
@@ -139,6 +139,16 @@
 {
 }
 
+#ifdef HAVE_POSIX_ACL
+static QString aclToText(acl_t acl) {
+    ssize_t size = 0;
+    char* txt = acl_to_text(acl, &size);
+    const QString ret = QString::fromLatin1(txt, size);
+    acl_free(txt);
+    return ret;
+}
+#endif
+
 int FileProtocol::setACL( const char *path, mode_t perm, bool directoryDefault )
 {
     int ret = 0;
@@ -157,8 +167,7 @@
         acl = acl_from_text( ACLString.toLatin1() );
         if ( acl_valid( acl ) == 0 ) { // let's be safe
             ret = acl_set_file( path, ACL_TYPE_ACCESS, acl );
-            ssize_t size = acl_size( acl );
-            kDebug(7101) << "Set ACL on: " << path << " to: " << acl_to_text( acl, &size );
+            kDebug(7101) << "Set ACL on: " << path << " to: " << aclToText(acl);
         }
         acl_free( acl );
         if ( ret != 0 ) return ret; // better stop trying right away
@@ -172,8 +181,7 @@
             acl_t acl = acl_from_text( defaultACLString.toLatin1() );
             if ( acl_valid( acl ) == 0 ) { // let's be safe
                 ret += acl_set_file( path, ACL_TYPE_DEFAULT, acl );
-                ssize_t size = acl_size( acl );
-                kDebug(7101) << "Set Default ACL on: " << path << " to: " << acl_to_text( acl, &size );
+                kDebug(7101) << "Set Default ACL on: " << path << " to: " << aclToText(acl);
             }
             acl_free( acl );
         }
@@ -1343,14 +1351,12 @@
     }
     if ( withACL ) {
         if ( acl ) {
-            ssize_t size = acl_size( acl );
-            const QString str = QString::fromLatin1( acl_to_text( acl, &size ) );
+            const QString str = aclToText(acl);
             entry.insert( KIO::UDSEntry::UDS_ACL_STRING, str );
             kDebug(7101) << path.data() << "ACL: " << str;
         }
         if ( defaultAcl ) {
-            ssize_t size = acl_size( defaultAcl );
-            const QString str = QString::fromLatin1( acl_to_text( defaultAcl, &size ) );
+            const QString str = aclToText(defaultAcl);
             entry.insert( KIO::UDSEntry::UDS_DEFAULT_ACL_STRING, str );
             kDebug(7101) << path.data() << "DEFAULT ACL: " << str;
         }
Index: kinit/klauncher.cpp
===================================================================
--- kinit/klauncher.cpp	(.../tags/KDE/4.2.1/kdelibs)	(wersja 935993)
+++ kinit/klauncher.cpp	(.../branches/KDE/4.2/kdelibs)	(wersja 935993)
@@ -971,9 +971,9 @@
        KService::Ptr service = KService::serviceByDesktopName(desktopName);
        if (service)
            send_service_startup_info(request, service,
-                                     request->startup_id, QStringList());
+                                     startup_id.toLocal8Bit(), QStringList());
        else // no .desktop file, no startup info
-           cancel_service_startup_info(request, request->startup_id, envs);
+           cancel_service_startup_info(request, startup_id.toLocal8Bit(), envs);
    }
    msg.setDelayedReply(true);
    request->transaction = msg;
@@ -1105,7 +1105,7 @@
     arg_list << protocol;
     arg_list << mConnectionServer.address();
     arg_list << app_socket;
-    name = KStandardDirs::findExe("kioslave");
+    name = KStandardDirs::findExe(QLatin1String("kioslave"));
 #else
     QString arg1 = protocol;
     QString arg2 = mConnectionServer.address();
Index: kio/kio/kfileitem.cpp
===================================================================
--- kio/kio/kfileitem.cpp	(.../tags/KDE/4.2.1/kdelibs)	(wersja 935993)
+++ kio/kio/kfileitem.cpp	(.../branches/KDE/4.2/kdelibs)	(wersja 935993)
@@ -420,10 +420,22 @@
     // even though it's not really part of the permissions per se.
     if (m_bLink)
         buffer[0] = 'l';
-    else if (m_fileMode != KFileItem::Unknown && S_ISDIR(m_fileMode))
-        buffer[0] = 'd';
-    else
+    else if (m_fileMode != KFileItem::Unknown) {
+        if (S_ISDIR(m_fileMode))
+            buffer[0] = 'd';
+        else if (S_ISSOCK(m_fileMode))
+            buffer[0] = 's';
+        else if (S_ISCHR(m_fileMode))
+            buffer[0] = 'c';
+        else if (S_ISBLK(m_fileMode))
+            buffer[0] = 'b';
+        else if (S_ISFIFO(m_fileMode))
+            buffer[0] = 'p';
+        else
+            buffer[0] = '-';
+    } else {
         buffer[0] = '-';
+    }
 
     buffer[1] = ((( perm & S_IRUSR ) == S_IRUSR ) ? 'r' : '-' );
     buffer[2] = ((( perm & S_IWUSR ) == S_IWUSR ) ? 'w' : '-' );
Index: kio/kio/kacl.cpp
===================================================================
--- kio/kio/kacl.cpp	(.../tags/KDE/4.2.1/kdelibs)	(wersja 935993)
+++ kio/kio/kacl.cpp	(.../branches/KDE/4.2/kdelibs)	(wersja 935993)
@@ -177,8 +177,9 @@
 #if 0
 static void printACL( acl_t acl, const QString &comment )
 {
-    ssize_t size = acl_size( acl );
-    kDebug() << comment << acl_to_text( acl, &size );
+    const char* txt = acl_to_text(acl);
+    kDebug() << comment << txt;
+    acl_free(txt);
 }
 #endif
 #endif
@@ -609,8 +610,11 @@
 QString KACL::asString() const
 {
 #ifdef HAVE_POSIX_ACL
-    ssize_t size = acl_size( d->m_acl );
-    return QString::fromLatin1( acl_to_text( d->m_acl, &size ) );
+    ssize_t size = 0;
+    char* txt = acl_to_text(d->m_acl, &size);
+    const QString ret = QString::fromLatin1(txt, size);
+    acl_free(txt);
+    return ret;
 #else
     return QString();
 #endif
Index: kio/kfile/kpropertiesdialog.cpp
===================================================================
--- kio/kfile/kpropertiesdialog.cpp	(.../tags/KDE/4.2.1/kdelibs)	(wersja 935993)
+++ kio/kfile/kpropertiesdialog.cpp	(.../branches/KDE/4.2/kdelibs)	(wersja 935993)
@@ -1732,7 +1732,6 @@
    */
   int i, maxEntries = 1000;
   struct passwd *user;
-  struct group *ge;
 
   /* File owner: For root, offer a KLineEdit with autocompletion.
    * For a user, who can never chown() a file, offer a QLabel.
@@ -1768,38 +1767,27 @@
     strUser = user->pw_name;
 
 #ifdef Q_OS_UNIX
-  setgrent();
-  for (i=0; ((ge = getgrent()) != 0L) && (i < maxEntries); ++i)
-  {
-    if (IamRoot)
-      groupList += QString::fromLatin1(ge->gr_name);
-    else
-    {
-      /* pick the groups to which the user belongs */
-      char ** members = ge->gr_mem;
-      char * member;
-      while ((member = *members) != 0L) {
-        if (strUser == member) {
-          groupList += QString::fromLocal8Bit(ge->gr_name);
-          break;
+    // pick the groups to which the user belongs
+    int groupCount = 0;
+    gid_t *groups = NULL;
+    if (getgrouplist(strUser, user->pw_gid, NULL, &groupCount) < 0) {
+        groups = new gid_t[groupCount];
+        if (groups) {
+            getgrouplist(strUser, user->pw_gid, groups, &groupCount);
+        } else {
+            groupCount = 0;
         }
-        ++members;
-      }
     }
-  }
-  endgrent();
-#endif //Q_OS_UNIX
 
-  /* add the effective Group to the list .. */
-  ge = getgrgid (getegid());
-  if (ge) {
-    QString name = QString::fromLatin1(ge->gr_name);
-    if (name.isEmpty())
-      name.setNum(ge->gr_gid);
-    if (groupList.indexOf(name) == -1)
-      groupList += name;
-  }
+    for (i = 0; i < groupCount; i++) {
+        struct group *mygroup = getgrgid(groups[i]);
+        if (mygroup)
+            groupList += QString::fromLocal8Bit(mygroup->gr_name);
+    }
 
+    delete[] groups;
+#endif //Q_OS_UNIX
+
   bool isMyGroup = groupList.contains(d->strGroup);
 
   /* add the group the file currently belongs to ..
Index: kio/misc/kpac/proxyscout.notifyrc
===================================================================
--- kio/misc/kpac/proxyscout.notifyrc	(.../tags/KDE/4.2.1/kdelibs)	(wersja 935993)
+++ kio/misc/kpac/proxyscout.notifyrc	(.../branches/KDE/4.2/kdelibs)	(wersja 935993)
@@ -334,7 +334,7 @@
 Comment[gu]=પ્રોક્સી રૂપરેખાંકન સ્ક્રિપ્ટ ડાઉનલોડ કરી શકાતી નથી
 Comment[he]=אין אפשרות להוריד את תסריט הגדרות שרת המתווך
 Comment[hi]=प्रॉक्सी कॉन्फ़िगरेशन स्क्रिप्ट को डाउनलोड नहीं किया जा सका
-Comment[hne]=प्राक्सी कान्फिगरेसन स्क्रिप्ट ल डाउनलोड नइ कर सकीस
+Comment[hne]=प्राक्सी कान्फिगरेसन स्क्रिप्ट ल डाउनलोड नइ कर सकिस
 Comment[hr]=Skripta konfiguriranja proxyja nije mogla biti preuzetom s Interneta
 Comment[hsb]=Konfiguraciski skript za proxy njehodźi so sćahnyć
 Comment[hu]=A proxyszkriptet nem sikerült letölteni
Index: kdeui/dialogs/kdialog.cpp
===================================================================
--- kdeui/dialogs/kdialog.cpp	(.../tags/KDE/4.2.1/kdelibs)	(wersja 935993)
+++ kdeui/dialogs/kdialog.cpp	(.../branches/KDE/4.2/kdelibs)	(wersja 935993)
@@ -149,6 +149,10 @@
   mButtonList.insert( key, button );
   mButtonSignalMapper.setMapping( button, key );
 
+  // since focus may not be on default button, disable auto-default
+  // (an auto-defaulted focused button "beats" the real default)
+  button->setAutoDefault( false );
+
     QObject::connect(button, SIGNAL(clicked()),
            &mButtonSignalMapper, SLOT( map() ) );
 }
Index: kdeui/widgets/ktextedit.cpp
===================================================================
--- kdeui/widgets/ktextedit.cpp	(.../tags/KDE/4.2.1/kdelibs)	(wersja 935993)
+++ kdeui/widgets/ktextedit.cpp	(.../branches/KDE/4.2/kdelibs)	(wersja 935993)
@@ -19,6 +19,7 @@
 */
 
 #include "ktextedit.h"
+#include <kdebug.h>
 
 #include <QApplication>
 #include <QClipboard>
@@ -364,6 +365,15 @@
     cursor.movePosition( QTextCursor::EndOfLine );
     parent->setTextCursor( cursor );
     return true;
+  } else if (KStandardShortcut::find().contains(key)) {
+        parent->slotFind();
+        return true;
+  } else if (KStandardShortcut::findNext().contains(key)) {
+        parent->slotFindNext();
+        return true;
+  } else if (KStandardShortcut::replace().contains(key)) {
+        parent->slotReplace();
+        return true;
   } else if ( KStandardShortcut::pasteSelection().contains( key ) ) {
     QString text = QApplication::clipboard()->text( QClipboard::Selection );
     if ( !text.isEmpty() )
@@ -407,7 +417,7 @@
           separatorAction = actionList.at( idx );
       if ( separatorAction )
       {
-          KAction *clearAllAction = KStandardAction::clear(this, SLOT(undoableClear()), this);
+          KAction *clearAllAction = KStandardAction::clear(this, SLOT(undoableClear()), popup);
           if ( emptyDocument )
               clearAllAction->setEnabled( false );
           popup->insertAction( separatorAction, clearAllAction );
@@ -434,9 +444,9 @@
 
       if (d->findReplaceEnabled)
       {
-          KAction *findAction = KStandardAction::find( this, SLOT( slotFind() ), this );
-          KAction *findNextAction = KStandardAction::findNext( this, SLOT( slotFindNext() ), this );
-          KAction *replaceAction = KStandardAction::replace( this, SLOT( slotReplace() ), this );
+          KAction *findAction = KStandardAction::find(this, SLOT(slotFind()), popup);
+          KAction *findNextAction = KStandardAction::findNext(this, SLOT(slotFindNext()), popup);
+          KAction *replaceAction = KStandardAction::replace(this, SLOT(slotReplace()), popup);
           if (emptyDocument)
           {
               findAction->setEnabled(false);
@@ -472,14 +482,14 @@
     QString selectedWord = wordSelectCursor.selectedText();
 
     // Clear the selection again, we re-select it below (without the apostrophes).
-    wordSelectCursor.movePosition(QTextCursor::PreviousCharacter,
-                                  QTextCursor::MoveAnchor, selectedWord.size());
+    wordSelectCursor.setPosition(wordSelectCursor.position()-selectedWord.size());
     if (selectedWord.startsWith('\'') || selectedWord.startsWith('\"')) {
         selectedWord = selectedWord.right(selectedWord.size() - 1);
         wordSelectCursor.movePosition(QTextCursor::NextCharacter, QTextCursor::MoveAnchor);
     }
     if (selectedWord.endsWith('\'') || selectedWord.endsWith('\"'))
         selectedWord.chop(1);
+
     wordSelectCursor.movePosition(QTextCursor::NextCharacter,
                                   QTextCursor::KeepAnchor, selectedWord.size());
 
@@ -909,6 +919,12 @@
     return true;
   } else if ( KStandardShortcut::pasteSelection().contains( key ) ) {
     return true;
+  } else if (KStandardShortcut::find().contains(key)) {
+      return true;
+  } else if (KStandardShortcut::findNext().contains(key)) {
+      return true;
+  } else if (KStandardShortcut::replace().contains(key)) {
+      return true;
   } else if (event->matches(QKeySequence::SelectAll)) { // currently missing in QTextEdit
       return true;
   } else if (event->modifiers() == Qt::ControlModifier &&
Index: kdeui/widgets/kkeysequencewidget.cpp
===================================================================
--- kdeui/widgets/kkeysequencewidget.cpp	(.../tags/KDE/4.2.1/kdelibs)	(wersja 935993)
+++ kdeui/widgets/kkeysequencewidget.cpp	(.../branches/KDE/4.2/kdelibs)	(wersja 935993)
@@ -248,7 +248,7 @@
     QString conflictingShortcuts;
     Q_FOREACH(const KAction *action, actions) {
         conflictingShortcuts += i18n("Shortcut(s) '%1' for action '%2'\n",
-                action->shortcut().toString(),
+                action->shortcut().toString(QKeySequence::NativeText),
                 KGlobal::locale()->removeAcceleratorMarker(action->text()));
     }
     QString message = i18n(
Index: kdeui/findreplace/kfind_p.h
===================================================================
--- kdeui/findreplace/kfind_p.h	(.../tags/KDE/4.2.1/kdelibs)	(wersja 935993)
+++ kdeui/findreplace/kfind_p.h	(.../branches/KDE/4.2/kdelibs)	(wersja 935993)
@@ -47,7 +47,9 @@
 
     ~Private()
     {
-        dialog->deleteLater();
+        if (dialog)
+            dialog->deleteLater();
+        dialog = 0;
         data.clear();
         delete emptyMatch;
         emptyMatch = 0;
Index: kdecore/kernel/kcmdlineargs.cpp
===================================================================
--- kdecore/kernel/kcmdlineargs.cpp	(.../tags/KDE/4.2.1/kdelibs)	(wersja 935993)
+++ kdecore/kernel/kcmdlineargs.cpp	(.../branches/KDE/4.2/kdelibs)	(wersja 935993)
@@ -83,13 +83,13 @@
 // Helper classes
 //
 
-class KCmdLineParsedOptions : public QHash<QString,QString>
+class KCmdLineParsedOptions : public QHash<QByteArray,QByteArray>
 {
 public:
    KCmdLineParsedOptions() { }
 };
 
-class KCmdLineParsedArgs : public QList<QString>
+class KCmdLineParsedArgs : public QList<QByteArray>
 {
 public:
    KCmdLineParsedArgs() { }
@@ -112,7 +112,7 @@
 
 class KCmdLineOptionsPrivate {
     public:
-    QStringList names;
+    QList<QByteArray> names;
     QList<KLocalizedString> descriptions;
     QStringList defaults;
 };
@@ -143,7 +143,7 @@
                                        const KLocalizedString &description,
                                        const QByteArray &defaultValue)
 {
-    d->names.append(QString::fromUtf8(name));
+    d->names.append(name);
     d->descriptions.append(description);
     d->defaults.append(QString::fromUtf8(defaultValue));
     return *this;
@@ -171,7 +171,7 @@
     char **argv; // The original argv
     bool parsed : 1; // Whether we have parsed the arguments since calling init
     bool ignoreUnknown : 1; // Ignore unknown options and arguments
-    QString mCwd; // Current working directory. Important for KUnqiueApp!
+    QByteArray mCwd; // Current working directory. Important for KUnqiueApp!
     KCmdLineArgs::StdCmdLineArgs mStdargs;
 
     KCmdLineOptions qt_options;
@@ -220,15 +220,15 @@
      *
      *  +4 - no more options follow         // !fork
      */
-    static int findOption(const KCmdLineOptions &options, QString &opt,
-                          QString &opt_name, QString &def, bool &enabled);
+    static int findOption(const KCmdLineOptions &options, QByteArray &opt,
+                          QByteArray &opt_name, QString &def, bool &enabled);
 
     /**
      * @internal
      *
      * Checks what to do with a single option
      */
-    static void findOption(const QString &optv, const QString &_opt,
+    static void findOption(const QByteArray &optv, const QByteArray &_opt,
                            int &i, bool _enabled, bool &moreOptions);
 
     /**
@@ -246,7 +246,7 @@
      *
      * @param id The name of the options to be removed.
      */
-    static void removeArgs(const QString &id);
+    static void removeArgs(const QByteArray &id);
 };
 
 K_GLOBAL_STATIC(KCmdLineArgsStatic, s)
@@ -336,7 +336,7 @@
 {
     friend class KCmdLineArgsStatic;
 public:
-    KCmdLineArgsPrivate(const KCmdLineOptions &_options, const KLocalizedString &_name, const QString &_id)
+    KCmdLineArgsPrivate(const KCmdLineOptions &_options, const KLocalizedString &_name, const QByteArray &_id)
         : options(_options)
         , name(_name)
         , id(_id)
@@ -351,7 +351,7 @@
     }
     const KCmdLineOptions options;
     const KLocalizedString name;
-    const QString id;
+    const QByteArray id;
     KCmdLineParsedOptions *parsedOptionList;
     KCmdLineParsedArgs *parsedArgList;
     bool isQt;
@@ -361,21 +361,21 @@
      *
      *  Set a boolean option
      */
-    void setOption(const QString &option, bool enabled);
+    void setOption(const QByteArray &option, bool enabled);
 
     /**
      * @internal
      *
      *  Set a string option
      */
-    void setOption(const QString &option, const QString &value);
+    void setOption(const QByteArray &option, const QByteArray &value);
 
     /**
      * @internal
      *
      * Add an argument
      */
-    void addArgument(const QString &argument);
+    void addArgument(const QByteArray &argument);
 
     /**
      * @internal
@@ -469,13 +469,13 @@
 
    s->about = _about;
    s->parsed = false;
-   s->mCwd = QDir::currentPath();
+   s->mCwd = QDir::currentPath().toLocal8Bit(); //currentPath() uses fromLocal8Bit internally apparently
    addStdCmdLineOptions(stdargs);
 }
 
 QString KCmdLineArgs::cwd()
 {
-   return s->mCwd;
+   return QString::fromLocal8Bit(s->mCwd);
 }
 
 QString KCmdLineArgs::appName()
@@ -499,14 +499,11 @@
 
 void
 KCmdLineArgs::addCmdLineOptions( const KCmdLineOptions &options, const KLocalizedString &name,
-         const QByteArray &_id, const QByteArray &_afterId)
+         const QByteArray &id, const QByteArray &afterId)
 {
    if (!s->argsList)
       s->argsList = new KCmdLineArgsList;
 
-   QString id = QString::fromUtf8(_id);
-   QString afterId = QString::fromUtf8(_afterId);
-
    int pos = s->argsList->count();
    // To make sure that the named options come before unnamed.
    if (pos > 0 && !id.isEmpty() && s->argsList->last()->d->name.isEmpty())
@@ -527,7 +524,7 @@
 
    Q_ASSERT( s->parsed == false ); // You must add _ALL_ cmd line options
                                    // before accessing the arguments!
-   s->argsList->insert(pos, new KCmdLineArgs(options, name, id.toUtf8()));
+   s->argsList->insert(pos, new KCmdLineArgs(options, name, id));
 }
 
 void
@@ -540,8 +537,7 @@
    s->removeArgs("qt");
    s->removeArgs("kde");
 
-   QByteArray qCwd = QFile::encodeName(s->mCwd);
-   ds << qCwd;
+   ds << s->mCwd;
 
    uint count = s->argsList ? s->argsList->count() : 0;
    ds << count;
@@ -579,16 +575,15 @@
    QByteArray qCwd;
    ds >> qCwd;
 
-   s->mCwd = QFile::decodeName(qCwd); //FIXME: Is this proper decoding?
+   s->mCwd = qCwd;
 
    uint count;
    ds >> count;
 
    while(count--)
    {
-     QByteArray idRaw;
-     ds >> idRaw;
-     QString id = idRaw;  //FIXME: What about decoding?
+     QByteArray id;
+     ds >> id;
      Q_ASSERT( s->argsList );
      for(args = s->argsList->begin(); args != s->argsList->end(); ++args)
      {
@@ -602,9 +597,8 @@
    s->parsed = true;
 }
 
-KCmdLineArgs *KCmdLineArgs::parsedArgs(const QByteArray &_id)
+KCmdLineArgs *KCmdLineArgs::parsedArgs(const QByteArray &id)
 {
-   QString id = QString::fromUtf8(_id);
    if (!s->argsList)
       return 0;
    KCmdLineArgsList::Iterator args = s->argsList->begin();
@@ -622,7 +616,7 @@
    return 0;
 }
 
-void KCmdLineArgsStatic::removeArgs(const QString &id)
+void KCmdLineArgsStatic::removeArgs(const QByteArray &id)
 {
    if (!s->argsList)
       return;
@@ -646,8 +640,8 @@
 }
 
 int
-KCmdLineArgsStatic::findOption(const KCmdLineOptions &options, QString &opt,
-                               QString &opt_name, QString &def, bool &enabled)
+KCmdLineArgsStatic::findOption(const KCmdLineOptions &options, QByteArray &opt,
+                               QByteArray &opt_name, QString &def, bool &enabled)
 {
    int result;
    bool inverse;
@@ -686,7 +680,7 @@
                i++;
                if (i >= options.d->names.size())
                   return result+0;
-               QString nextOption = options.d->names[i];
+               QByteArray nextOption = options.d->names[i];
                int p = nextOption.indexOf(' ');
                if (p > 0)
                   nextOption = nextOption.left(p);
@@ -717,14 +711,14 @@
 }
 
 void
-KCmdLineArgsStatic::findOption(const QString &optv, const QString &_opt,
+KCmdLineArgsStatic::findOption(const QByteArray &optv, const QByteArray &_opt,
                                int &i, bool _enabled, bool &moreOptions)
 {
    KCmdLineArgsList::Iterator args = s->argsList->begin();
-   QString opt = _opt;
-   QString opt_name;
+   QByteArray opt = _opt;
+   QByteArray opt_name;
    QString def;
-   QString argument;
+   QByteArray argument;
    int j = opt.indexOf('=');
    if (j != -1)
    {
@@ -753,7 +747,7 @@
       int p = 1;
       while (true)
       {
-         QString singleCharOption = " ";
+         QByteArray singleCharOption = " ";
          singleCharOption[0] = optv[p];
          args = s->argsList->begin();
          while (args != s->argsList->end())
@@ -800,7 +794,7 @@
 			return;
 #endif
       KCmdLineArgs::enable_i18n();
-      KCmdLineArgs::usageError( i18n("Unknown option '%1'.", _opt));
+      KCmdLineArgs::usageError( i18n("Unknown option '%1'.", QString::fromLocal8Bit(_opt)));
    }
 
    if ((result & 4) != 0)
@@ -820,7 +814,7 @@
 				return;
 #endif
          KCmdLineArgs::enable_i18n();
-         KCmdLineArgs::usageError( i18n("Unknown option '%1'.", _opt));
+         KCmdLineArgs::usageError( i18n("Unknown option '%1'.", QString::fromLocal8Bit(_opt)));
       }
       if (argument.isEmpty())
       {
@@ -828,9 +822,9 @@
          if (i >= s->argc)
          {
             KCmdLineArgs::enable_i18n();
-            KCmdLineArgs::usageError( i18nc("@info:shell %1 is cmdoption name","'%1' missing.",  opt_name));
+            KCmdLineArgs::usageError( i18nc("@info:shell %1 is cmdoption name","'%1' missing.",  QString::fromLocal8Bit(opt_name)));
          }
-         argument = s->decodeInput(s->argv[i]);
+         argument = s->argv[i];
       }
       (*args)->d->setOption(opt, argument);
    }
@@ -849,16 +843,10 @@
    KCmdLineArgs *appOptions = s->argsList->last();
    if (appOptions->d->id.isEmpty())
    {
-     const KCmdLineOptions &option = appOptions->d->options;
-     for (int i = 0; i < option.d->names.size(); i++)
+     foreach(const QByteArray& name, appOptions->d->options.d->names)
      {
-       if (option.d->names[i].startsWith('+'))
-           allowArgs = true;
-       if ( option.d->names[i].startsWith("!+") )
-       {
-           allowArgs = true;
-           everythingAfterArgIsArgs = true;
-       }
+       everythingAfterArgIsArgs = everythingAfterArgIsArgs || name.startsWith("!+");
+       allowArgs = allowArgs || name.startsWith('+') || everythingAfterArgIsArgs;
      }
    }
    for(int i = 1; i < s->argc; i++)
@@ -869,8 +857,8 @@
       if ((s->argv[i][0] == '-') && s->argv[i][1] && inOptions)
       {
          bool enabled = true;
-         QString orig = decodeInput(s->argv[i]);
-         QString option = orig.mid(1);
+         QByteArray orig = s->argv[i];
+         QByteArray option = orig.mid(1);
          if (option.startsWith('-'))
          {
             option = option.mid(1);
@@ -887,7 +875,7 @@
          }
          else if (option.startsWith("help-"))
          {
-            KCmdLineArgs::usage(option.mid(5).toUtf8());
+            KCmdLineArgs::usage(option.mid(5));
          }
          else if ((option == "version") || (option == "v"))
          {
@@ -924,9 +912,8 @@
          {
            if (s->about->bugAddress().isEmpty() || s->about->bugAddress() == "submit@bugs.kde.org" )
              s->printQ( i18n( "Please use http://bugs.kde.org to report bugs.\n" ) );
-           else {
+           else
              s->printQ( i18n( "Please report bugs to %1.\n" , s->about->bugAddress()) );
-           }
          }
          else
          {
@@ -955,7 +942,7 @@
          }
          else
          {
-            appOptions->d->addArgument(s->decodeInput(s->argv[i]));
+            appOptions->d->addArgument(s->argv[i]);
             if (everythingAfterArgIsArgs)
                 inOptions = false;
          }
@@ -1009,7 +996,7 @@
    if (!(s->mStdargs & KCmdLineArgs::CmdLineArgQt))
    {
      s_qt_argv = new char*[2];
-     s_qt_argv[0] = qstrdup(s->encodeOutput(appName()));
+     s_qt_argv[0] = qstrdup(s->argc?s->argv[0]:"");
      s_qt_argv[1] = 0;
 
      return s_qt_argv;
@@ -1026,12 +1013,13 @@
       exit(255);
    }
 
-   s_qt_argv = new char*[ args->count() + 2 ];
-   s_qt_argv[0] = qstrdup(s->encodeOutput(appName()));
+   int count=args->count();
+   s_qt_argv = new char*[ count + 2 ];
+   s_qt_argv[0] = qstrdup(s->argc?s->argv[0]:"");
    int i = 0;
-   for(; i < args->count(); i++)
+   for(; i < count; i++)
    {
-      s_qt_argv[i+1] = qstrdup(s->encodeOutput(args->arg(i)));
+      s_qt_argv[i+1] = qstrdup(args->d->parsedArgList->at(i));
    }
    s_qt_argv[i+1] = 0;
 
@@ -1110,7 +1098,7 @@
      const KCmdLineOptions &option = appOptions->d->options;
      for (int i = 0; i < option.d->names.size(); i++)
      {
-       QString opt_name = option.d->names[i];
+       QByteArray opt_name = option.d->names[i];
        if (opt_name.startsWith('+'))
           usage = usage + (opt_name.mid(1)) + ' ';
        else if ( opt_name.startsWith("!+") )
@@ -1129,7 +1117,7 @@
    {
       if (!(*args)->d->name.isEmpty() && !(*args)->d->id.isEmpty())
       {
-         QString option = QString("--help-%1").arg((*args)->d->id);
+         QString option = QString("--help-%1").arg(QString::fromLatin1((*args)->d->id));
          QString desc = i18n("Show %1 specific options", (*args)->d->name.toString());
 
          s->printQ(optionFormatString.arg(option, -25).arg(desc));
@@ -1169,7 +1157,7 @@
      while (args != s->argsList->end())
      {
        const KCmdLineOptions &option = (*args)->d->options;
-       QString opt;
+       QByteArray opt;
 
        for (int i = 0; i < option.d->names.size(); i++)
        {
@@ -1214,7 +1202,7 @@
             description = dl.first();
             dl.erase( dl.begin() );
          }
-         QString name = option.d->names[i];
+         QByteArray name = option.d->names[i];
          if (name.startsWith('!'))
              name = name.mid(1);
 
@@ -1229,7 +1217,7 @@
             name = name.mid(1);
             if (name.startsWith('[') && name.endsWith(']'))
                 name = name.mid(1, name.length()-2);
-            s->printQ(optionFormatString.arg(name, -25).arg(description));
+            s->printQ(optionFormatString.arg(QString::fromLocal8Bit(name), -25).arg(description));
          }
          else
          {
@@ -1259,7 +1247,7 @@
                   s->printQ(optionFormatStringDef.arg(QString( opt ), -25)
                             .arg(description, option.d->defaults[i]));
                }
-               opt = "";
+               opt.clear();
             }
          }
          for(QStringList::Iterator it = dl.begin();
@@ -1309,24 +1297,21 @@
 void
 KCmdLineArgs::setCwd( const QByteArray &cwd )
 {
-    s->mCwd = QString::fromUtf8(cwd);
+    s->mCwd = cwd;
 }
 
 void
 KCmdLineArgs::clear()
 {
-   delete d->parsedArgList;
-   d->parsedArgList = 0;
-   delete d->parsedOptionList;
-   d->parsedOptionList = 0;
+   delete d->parsedArgList;     d->parsedArgList = 0;
+   delete d->parsedOptionList;  d->parsedOptionList = 0;
 }
 
 void
 KCmdLineArgs::reset()
 {
    if ( s->argsList ) {
-      delete s->argsList;
-      s->argsList = 0;
+      delete s->argsList; s->argsList = 0;
    }
    s->parsed = false;
 }
@@ -1356,23 +1341,21 @@
 
    if (parsedOptionList->count() == 0)
    {
-      delete parsedOptionList;
-      parsedOptionList = 0;
+      delete parsedOptionList;  parsedOptionList = 0;
    }
    if (parsedArgList->count() == 0)
    {
-      delete parsedArgList;
-      parsedArgList = 0;
+      delete parsedArgList;     parsedArgList = 0;
    }
 }
 
 void
-KCmdLineArgsPrivate::setOption(const QString &opt, bool enabled)
+KCmdLineArgsPrivate::setOption(const QByteArray &opt, bool enabled)
 {
    if (isQt)
    {
       // Qt does it own parsing.
-      QString argString = "-";
+      QByteArray argString = "-";
       if( !enabled )
           argString += "no";
       argString += opt;
@@ -1383,18 +1366,18 @@
    }
 
    if (enabled)
-      parsedOptionList->insert( opt, QString::fromUtf8("t") );
+      parsedOptionList->insert( opt, "t" );
    else
-      parsedOptionList->insert( opt, QString::fromUtf8("f") );
+      parsedOptionList->insert( opt, "f" );
 }
 
 void
-KCmdLineArgsPrivate::setOption(const QString &opt, const QString &value)
+KCmdLineArgsPrivate::setOption(const QByteArray &opt, const QByteArray &value)
 {
    if (isQt)
    {
       // Qt does it's own parsing.
-      QString argString = "-";
+      QByteArray argString = "-";
       argString += opt;
       addArgument(argString);
       addArgument(value);
@@ -1403,7 +1386,7 @@
       // Hack coming up!
       if (argString == "-display")
       {
-         setenv(DISPLAY, s->encodeOutput(value).data(), true);
+         setenv(DISPLAY, value.data(), true);
       }
 #endif
    }
@@ -1417,8 +1400,8 @@
 QString
 KCmdLineArgs::getOption(const QByteArray &_opt) const
 {
-   QString opt = QString::fromUtf8(_opt);
-   QString value;
+   QByteArray opt = _opt;
+   QByteArray value;
    if (d->parsedOptionList)
    {
       value = d->parsedOptionList->value(opt);
@@ -1427,7 +1410,7 @@
       return value;
 
    // Look up the default.
-   QString opt_name;
+   QByteArray opt_name;
    QString def;
    bool dummy = true;
    int result = s->findOption( d->options, opt, opt_name, def, dummy) & ~4;
@@ -1436,7 +1419,7 @@
    {
       fprintf(stderr, "\n\nFAILURE (KCmdLineArgs):\n");
       fprintf(stderr, "Application requests for getOption(\"%s\") but the \"%s\" option\n",
-                      s->encodeOutput(opt).data(), s->encodeOutput(opt).data());
+                      opt.data(), opt.data());
       fprintf(stderr, "has never been specified via addCmdLineOptions( ... )\n\n");
 
       Q_ASSERT( 0 );
@@ -1446,20 +1429,18 @@
 }
 
 QStringList
-KCmdLineArgs::getOptionList(const QByteArray &_opt) const
+KCmdLineArgs::getOptionList(const QByteArray &opt) const
 {
-   QString opt = QString::fromUtf8(_opt);
-
    QStringList result;
    if (!d->parsedOptionList)
       return result;
 
    while(true)
    {
-      QString value = d->parsedOptionList->take(opt);
+      QByteArray value = d->parsedOptionList->take(opt);
       if (value.isEmpty())
          break;
-      result.prepend(value);
+      result.prepend(QString::fromLocal8Bit(value));
    }
 
    // Reinsert items in dictionary
@@ -1469,7 +1450,7 @@
    // So taking them out and then putting them back is the only way.
    Q_FOREACH(const QString &str, result)
    {
-      d->parsedOptionList->insertMulti(opt, str);
+      d->parsedOptionList->insertMulti(opt, str.toLocal8Bit());
    }
    return result;
 }
@@ -1478,8 +1459,8 @@
 KCmdLineArgs::isSet(const QByteArray &_opt) const
 {
    // Look up the default.
-   QString opt = QString::fromUtf8(_opt);
-   QString opt_name;
+   QByteArray opt = _opt;
+   QByteArray opt_name;
    QString def;
    int result = 0;
    KCmdLineArgsList::Iterator args = s->argsList->begin();
@@ -1495,14 +1476,14 @@
    {
       fprintf(stderr, "\n\nFAILURE (KCmdLineArgs):\n");
       fprintf(stderr, "Application requests for isSet(\"%s\") but the \"%s\" option\n",
-                      s->encodeOutput(opt).data(), s->encodeOutput(opt).data());
+                      opt.data(), opt.data());
       fprintf(stderr, "has never been specified via addCmdLineOptions( ... )\n\n");
 
       Q_ASSERT( 0 );
       exit(255);
    }
 
-   QString value;
+   QByteArray value;
    if (d->parsedOptionList)
    {
       value = d->parsedOptionList->value(opt);
@@ -1513,7 +1494,7 @@
       if (result == 3)
          return true;
       else
-         return (value[0] == 't');
+         return (value.at(0) == 't');
    }
 
    if (result == 3)
@@ -1527,9 +1508,7 @@
 int
 KCmdLineArgs::count() const
 {
-   if (!d->parsedArgList)
-      return 0;
-   return d->parsedArgList->count();
+   return d->parsedArgList?d->parsedArgList->count():0;
 }
 
 QString
@@ -1545,7 +1524,7 @@
       exit(255);
    }
 
-   return d->parsedArgList->at(n);
+   return QString::fromLocal8Bit(d->parsedArgList->at(n));
 }
 
 KUrl
@@ -1575,7 +1554,7 @@
 }
 
 void
-KCmdLineArgsPrivate::addArgument(const QString &argument)
+KCmdLineArgsPrivate::addArgument(const QByteArray &argument)
 {
    if (!parsedArgList)
       parsedArgList = new KCmdLineParsedArgs;
@@ -1594,7 +1573,5 @@
 bool KCmdLineArgs::isTempFileSet()
 {
     KCmdLineArgs* args = KCmdLineArgs::parsedArgs( "kde-tempfile" );
-    if ( args )
-        return args->isSet( "tempfile" );
-    return false;
+    return args && args->isSet( "tempfile" );
 }
Index: kdecore/services/yacc.y
===================================================================
--- kdecore/services/yacc.y	(.../tags/KDE/4.2.1/kdelibs)	(wersja 935993)
+++ kdecore/services/yacc.y	(.../branches/KDE/4.2/kdelibs)	(wersja 935993)
@@ -137,4 +137,5 @@
 {
   KTraderParse_initFlex( _code );
   yyparse();
+  kiotraderlex_destroy();
 }
Index: kdecore/services/kservicetypeprofile.cpp
===================================================================
--- kdecore/services/kservicetypeprofile.cpp	(.../tags/KDE/4.2.1/kdelibs)	(wersja 935993)
+++ kdecore/services/kservicetypeprofile.cpp	(.../branches/KDE/4.2/kdelibs)	(wersja 935993)
@@ -84,7 +84,7 @@
             const QString serviceId = config.readEntry( "Entry" + num + "_Service", QString() );
             Q_ASSERT(!serviceId.isEmpty());
             const int pref = config.readEntry( "Entry" + num + "_Preference", 0 );
-            //kDebug(7014) << "KServiceTypeProfile::initStatic adding service " << serviceId << " to profile for " << type << " with preference " << pref;
+            //kDebug(7014) << "adding service " << serviceId << " to profile for " << type << " with preference " << pref;
             p->addService( serviceId, pref );
         }
     }
@@ -210,8 +210,9 @@
     config.deleteGroup( serviceType );
     config.sync();
 
-    if (s_serviceTypeProfiles.exists())
-        s_serviceTypeProfiles->remove( serviceType );
+    if (s_serviceTypeProfiles.exists()) {
+        delete s_serviceTypeProfiles->take( serviceType );
+    }
 }
 
 void KServiceTypeProfile::setConfigurationMode()
Index: kdecore/all_languages.desktop
===================================================================
--- kdecore/all_languages.desktop	(.../tags/KDE/4.2.1/kdelibs)	(wersja 935993)
+++ kdecore/all_languages.desktop	(.../branches/KDE/4.2/kdelibs)	(wersja 935993)
@@ -4154,6 +4154,9 @@
 Name[ca]=Chattisgarbi
 Name[es]=Chhattisgarhí
 Name[hu]=Cshattíszgarhi
+Name[ja]=チャッティースガリー語
+Name[ru]=Чхаттисгархи
+Name[se]=Čattisgarhigiella
 Name[sr]=чатисгархи
 Name[sr@latin]=чатисгархи
 Name[uk]=Чхатісгархі
Index: kdecore/sonnet/sonnetspeller.desktop
===================================================================
--- kdecore/sonnet/sonnetspeller.desktop	(.../tags/KDE/4.2.1/kdelibs)	(wersja 935993)
+++ kdecore/sonnet/sonnetspeller.desktop	(.../branches/KDE/4.2/kdelibs)	(wersja 935993)
@@ -45,6 +45,7 @@
 Comment[mk]=Sonnet-клиент за правопис
 Comment[ml]=സോണറ്റ് സ്പെല്‍ ക്ലയന്റ്
 Comment[mr]=सोनेट शब्दलेखन क्लाएंट
+Comment[ms]=Klien Ejaan Sonnet
 Comment[nb]=Sonnet staveklient
 Comment[nds]=Schriefwiesprööv-Client Sonnet
 Comment[ne]=सोनेट स्पेल क्लाइन्ट
Index: kdecore/kconfig_compiler/kconfig_compiler.cpp
===================================================================
--- kdecore/kconfig_compiler/kconfig_compiler.cpp	(.../tags/KDE/4.2.1/kdelibs)	(wersja 935993)
+++ kdecore/kconfig_compiler/kconfig_compiler.cpp	(.../branches/KDE/4.2/kdelibs)	(wersja 935993)
@@ -442,8 +442,9 @@
 
     } else if ( type == "Path" && !defaultValue.isEmpty() ) {
       defaultValue = literalString( defaultValue );
-
-    } else if ( (type == "StringList" || type == "PathList") && !defaultValue.isEmpty() ) {
+    } else if ( type == "Url" && !defaultValue.isEmpty() ) {
+      defaultValue = "KUrl( " + literalString(defaultValue) + ")";
+    } else if ( ( type == "UrlList" || type == "StringList" || type == "PathList") && !defaultValue.isEmpty() ) {
       QTextStream cpp( &code, QIODevice::WriteOnly | QIODevice::Append );
       if (!code.isEmpty())
          cpp << endl;
@@ -452,8 +453,15 @@
       const QStringList defaults = defaultValue.split( ',' );
       QStringList::ConstIterator it;
       for( it = defaults.constBegin(); it != defaults.constEnd(); ++it ) {
-        cpp << "  default" << name << ".append( QString::fromUtf8( \"" << *it << "\" ) );"
-            << endl;
+        cpp << "  default" << name << ".append( ";
+        if( type == "UrlList" ) {
+          cpp << "KUrl(";       
+        }
+        cpp << "QString::fromUtf8( \"" << *it << "\" ) ";
+        if( type == "UrlList" ) {
+          cpp << ") ";
+        }
+        cpp << ");" << endl;
       }
       defaultValue = "default" + name;
 
Index: plasma/servicetypes/plasma-containment.desktop
===================================================================
--- plasma/servicetypes/plasma-containment.desktop	(.../tags/KDE/4.2.1/kdelibs)	(wersja 935993)
+++ plasma/servicetypes/plasma-containment.desktop	(.../branches/KDE/4.2/kdelibs)	(wersja 935993)
@@ -20,7 +20,7 @@
 Comment[gl]=Un contedor de applet de Plasma e pintor do fondo
 Comment[gu]=પ્લાઝમા એપ્લેટ ધરાવનાર અને પાશ્ચભાગ રંગનાર
 Comment[he]=תוחם של יישומון Plasma וצובע הרקע
-Comment[hne]=प्लाज्मा एपलेट कंटेनर अउ पिछोत पुतइया
+Comment[hne]=प्लाज्मा एपलेट कंटेनर अउ पिछोत अंगना पुतइया
 Comment[hsb]=Plasma-container za applet a molowadło pozadka
 Comment[hu]=Tartóelem és háttérrajzoló Plasma-kisalkalmazásokhoz
 Comment[is]=Grunnur fyrir Plasma smáforrit og bakgrunnslitun
Index: plasma/servicetypes/plasma-applet-extenderapplet.desktop
===================================================================
--- plasma/servicetypes/plasma-applet-extenderapplet.desktop	(.../tags/KDE/4.2.1/kdelibs)	(wersja 935993)
+++ plasma/servicetypes/plasma-applet-extenderapplet.desktop	(.../branches/KDE/4.2/kdelibs)	(wersja 935993)
@@ -50,6 +50,7 @@
 Name[zh_TW]=內部延伸容器
 Type=Service
 Icon=utilities-desktop-extra
+NoDisplay=true
 X-KDE-ServiceTypes=Plasma/Applet
 X-KDE-PluginInfo-Name=internal:extender
 X-KDE-PluginInfo-EnabledByDefault=false
Index: plasma/abstractrunner.cpp
===================================================================
--- plasma/abstractrunner.cpp	(.../tags/KDE/4.2.1/kdelibs)	(wersja 935993)
+++ plasma/abstractrunner.cpp	(.../branches/KDE/4.2/kdelibs)	(wersja 935993)
@@ -132,14 +132,13 @@
 {
 }
 
-void AbstractRunner::performMatch(Plasma::RunnerContext &globalContext)
+void AbstractRunner::performMatch(Plasma::RunnerContext &localContext)
 {
     static const int reasonableRunTime = 1500;
     static const int fastEnoughTime = 250;
 
     d->runtime.restart();
-//TODO :this is a copy ctor
-    RunnerContext localContext(globalContext, 0);
+    
     match(localContext);
     // automatically rate limit runners that become slooow
     const int runtime = d->runtime.elapsed();
Index: plasma/runnercontext.cpp
===================================================================
--- plasma/runnercontext.cpp	(.../tags/KDE/4.2.1/kdelibs)	(wersja 935993)
+++ plasma/runnercontext.cpp	(.../branches/KDE/4.2/kdelibs)	(wersja 935993)
@@ -142,8 +142,15 @@
 {
     // We will detach if we are a copy of someone. But we will reset
     // if we are the 'main' context others copied from. Resetting
-    // one RunnerContext makes all the copies oneobsolete.
+    // one RunnerContext makes all the copies obsolete.
+    
+    d->q = 0; //we need to mark the q pointer of the detached RunnerContextPrivate 
+              //as dirty on detach to avoid receiving results for old queries   
+
     d.detach();
+    
+    d->q = this; //now that we detached the d pointer we need to mark its q pointer as
+                 //this to receive the signals 
 
     // we still have to remove all the matches, since if the
     // ref count was 1 (e.g. only the RunnerContext is using
@@ -151,7 +158,7 @@
     if (!d->matches.isEmpty()) {
         d->matchesById.clear();
         d->matches.clear();
-        emit d->q->matchesChanged();
+        emit matchesChanged();
     }
 
     d->term.clear();
@@ -194,7 +201,7 @@
 {
     Q_UNUSED(term)
 
-    if (matches.isEmpty()) {
+    if (matches.isEmpty() || (!d->q)) { //Bail out if the query is empty or the qptr is dirty 
         return false;
     }
 
@@ -214,6 +221,7 @@
     // we always want to sent the signal of the object that created
     // the d pointer
     emit d->q->matchesChanged();
+    
     return true;
 }
 
@@ -221,12 +229,17 @@
 {
     Q_UNUSED(term)
 
+    if (!d->q) { // Bail out if the qptr is dirty
+        return false;
+    }
+    
     LOCK_FOR_WRITE(this)
     d->matches.append(match);
     d->matchesById.insert(match.id(), &d->matches.at(d->matches.size() - 1));
     UNLOCK(this);
     //kDebug()<< "added match" << match->text();
-    emit d->q->matchesChanged();
+    emit d->q->matchesChanged(); 
+    
 
     return true;
 }
Index: plasma/runnermanager.cpp
===================================================================
--- plasma/runnermanager.cpp	(.../tags/KDE/4.2.1/kdelibs)	(wersja 935993)
+++ plasma/runnermanager.cpp	(.../branches/KDE/4.2/kdelibs)	(wersja 935993)
@@ -138,19 +138,23 @@
 
     int priority() const;
     Plasma::AbstractRunner *runner() const;
+    void setStale();
+    bool isStale() const;
 
 protected:
     void run();
 private:
-    Plasma::RunnerContext *m_context;
+    Plasma::RunnerContext m_context;
     Plasma::AbstractRunner *m_runner;
+    bool m_stale;
 };
 
 FindMatchesJob::FindMatchesJob(Plasma::AbstractRunner *runner,
                                Plasma::RunnerContext *context, QObject *parent)
     : ThreadWeaver::Job(parent),
-      m_context(context),
-      m_runner(runner)
+      m_context(*context, 0),
+      m_runner(runner),
+      m_stale(false)
 {
     if (runner->speed() == Plasma::AbstractRunner::SlowSpeed) {
         assignQueuePolicy(&RunnerRestrictionPolicy::instance());
@@ -161,7 +165,7 @@
 {
 //     kDebug() << "Running match for " << m_runner->objectName()
 //              << " in Thread " << thread()->id() << endl;
-    m_runner->performMatch(*m_context);
+    m_runner->performMatch(m_context);
 }
 
 int FindMatchesJob::priority() const
@@ -174,6 +178,16 @@
     return m_runner;
 }
 
+void FindMatchesJob::setStale() 
+{
+    m_stale = true;
+}
+
+bool FindMatchesJob::isStale() const
+{
+    return m_stale;
+}
+
 /*****************************************************
 *  RunnerManager::Private class
 *
@@ -287,8 +301,9 @@
         FindMatchesJob *runJob = static_cast<FindMatchesJob*>(job);
         if (deferredRun.isEnabled() && runJob->runner() == deferredRun.runner()) {
             //kDebug() << "job actually done, running now **************";
-            deferredRun.run(context);
-            deferredRun = QueryMatch(0);
+            QueryMatch tmpRun = deferredRun;
+            deferredRun = QueryMatch(0);	  
+            tmpRun.run(context);
         }
         searchJobs.removeAll(runJob);
         delete runJob;
@@ -375,18 +390,21 @@
     AbstractRunner *runner = match.runner();
 
     foreach (FindMatchesJob *job, d->searchJobs) {
-        if (job->runner() == runner && !job->isFinished()) {
-            //kDebug() << "!!!!!!!!!!!!!!!!!!! uh oh!";
+        if (job->runner() == runner && !job->isFinished() && !job->isStale()) {
+            kDebug() << "!!!!!!!!!!!!!!!!!!! uh oh!";
             d->deferredRun = match;
             return;
         }
     }
 
-    match.run(d->context);
-
     if (d->deferredRun.isValid()) {
         d->deferredRun = QueryMatch(0);
     }
+    
+    match.run(d->context);
+
+
+    
 }
 
 QList<QAction*> RunnerManager::actionsForMatch(const QueryMatch &match)
@@ -483,8 +501,8 @@
         //kDebug() << "ignored!";
         return false;
     }
-
-    r->performMatch(d->context);
+    RunnerContext localContext(d->context, 0);
+    r->performMatch(localContext);
     //kDebug() << "succeeded with" << d->context.matches().count() << "results";
     emit matchesChanged(d->context.matches());
     return true;
@@ -503,12 +521,18 @@
         d->searchJobs.clear();
     } else {
         Weaver::instance()->dequeue();
+        // Since we cannot safely delete the jobs, mark them as stale 
+        // TODO: delete them eventually?
+        foreach (FindMatchesJob *job, d->searchJobs) {
+	    job->setStale();
+        }
     }
 
     if (d->deferredRun.isEnabled()) {
         //kDebug() << "job actually done, running now **************";
-        d->deferredRun.run(d->context);
+        QueryMatch tmpRun = d->deferredRun;
         d->deferredRun = QueryMatch(0);
+        tmpRun.run(d->context);
     }
 
     d->context.reset();

Zmiany atrybutów dla: .
___________________________________________________________________
Dodane: svn:externals
   + 


